<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Dimension reduction of the transcript abundance data — reduce_dimensions • tidybulk</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Dimension reduction of the transcript abundance data — reduce_dimensions" />
<meta property="og:description" content="reduce_dimensions() takes as input a `tbl` formatted as | &amp;lt;SAMPLE&amp;gt; | &amp;lt;TRANSCRIPT&amp;gt; | &amp;lt;COUNT&amp;gt; | &amp;lt;...&amp;gt; | and calculates the reduced dimensional space of the transcript abundance." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidybulk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/comparison_with_base_R.html">Comparison with base R</a>
    </li>
    <li>
      <a href="../articles/introduction.html">Overview of the tidybulk package</a>
    </li>
    <li>
      <a href="../articles/manuscript_differential_transcript_abundance.html">Manuscript code - differential transcript abundance</a>
    </li>
    <li>
      <a href="../articles/manuscript_transcriptional_signatures.html">Manuscript code - transcriptional signature identification</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Dimension reduction of the transcript abundance data</h1>
    
    <div class="hidden name"><code>reduce_dimensions-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>reduce_dimensions() takes as input a `tbl` formatted as | &lt;SAMPLE&gt; | &lt;TRANSCRIPT&gt; | &lt;COUNT&gt; | &lt;...&gt; | and calculates the reduced dimensional space of the transcript abundance.</p>
    </div>

    <pre class="usage"><span class='fu'>reduce_dimensions</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  .element <span class='op'>=</span> <span class='cn'>NULL</span>,
  .feature <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>method</span>,
  .dims <span class='op'>=</span> <span class='fl'>2</span>,
  top <span class='op'>=</span> <span class='fl'>500</span>,
  of_samples <span class='op'>=</span> <span class='cn'>TRUE</span>,
  log_transform <span class='op'>=</span> <span class='cn'>TRUE</span>,
  scale <span class='op'>=</span> <span class='cn'>TRUE</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='co'># S4 method for spec_tbl_df</span>
<span class='fu'>reduce_dimensions</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  .element <span class='op'>=</span> <span class='cn'>NULL</span>,
  .feature <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>method</span>,
  .dims <span class='op'>=</span> <span class='fl'>2</span>,
  top <span class='op'>=</span> <span class='fl'>500</span>,
  of_samples <span class='op'>=</span> <span class='cn'>TRUE</span>,
  log_transform <span class='op'>=</span> <span class='cn'>TRUE</span>,
  scale <span class='op'>=</span> <span class='cn'>TRUE</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='co'># S4 method for tbl_df</span>
<span class='fu'>reduce_dimensions</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  .element <span class='op'>=</span> <span class='cn'>NULL</span>,
  .feature <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>method</span>,
  .dims <span class='op'>=</span> <span class='fl'>2</span>,
  top <span class='op'>=</span> <span class='fl'>500</span>,
  of_samples <span class='op'>=</span> <span class='cn'>TRUE</span>,
  log_transform <span class='op'>=</span> <span class='cn'>TRUE</span>,
  scale <span class='op'>=</span> <span class='cn'>TRUE</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='co'># S4 method for tidybulk</span>
<span class='fu'>reduce_dimensions</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  .element <span class='op'>=</span> <span class='cn'>NULL</span>,
  .feature <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>method</span>,
  .dims <span class='op'>=</span> <span class='fl'>2</span>,
  top <span class='op'>=</span> <span class='fl'>500</span>,
  of_samples <span class='op'>=</span> <span class='cn'>TRUE</span>,
  log_transform <span class='op'>=</span> <span class='cn'>TRUE</span>,
  scale <span class='op'>=</span> <span class='cn'>TRUE</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='co'># S4 method for SummarizedExperiment</span>
<span class='fu'>reduce_dimensions</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  .element <span class='op'>=</span> <span class='cn'>NULL</span>,
  .feature <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>method</span>,
  .dims <span class='op'>=</span> <span class='fl'>2</span>,
  top <span class='op'>=</span> <span class='fl'>500</span>,
  of_samples <span class='op'>=</span> <span class='cn'>TRUE</span>,
  log_transform <span class='op'>=</span> <span class='cn'>TRUE</span>,
  scale <span class='op'>=</span> <span class='cn'>TRUE</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='co'># S4 method for RangedSummarizedExperiment</span>
<span class='fu'>reduce_dimensions</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  .element <span class='op'>=</span> <span class='cn'>NULL</span>,
  .feature <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>method</span>,
  .dims <span class='op'>=</span> <span class='fl'>2</span>,
  top <span class='op'>=</span> <span class='fl'>500</span>,
  of_samples <span class='op'>=</span> <span class='cn'>TRUE</span>,
  log_transform <span class='op'>=</span> <span class='cn'>TRUE</span>,
  scale <span class='op'>=</span> <span class='cn'>TRUE</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>.data</th>
      <td><p>A `tbl` formatted as | &lt;SAMPLE&gt; | &lt;TRANSCRIPT&gt; | &lt;COUNT&gt; | &lt;...&gt; |</p></td>
    </tr>
    <tr>
      <th>.element</th>
      <td><p>The name of the element column (normally samples).</p></td>
    </tr>
    <tr>
      <th>.feature</th>
      <td><p>The name of the feature column (normally transcripts/genes)</p></td>
    </tr>
    <tr>
      <th>.abundance</th>
      <td><p>The name of the column including the numerical value the clustering is based on (normally transcript abundance)</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>A character string. The dimension reduction algorithm to use (PCA, MDS, tSNE).</p></td>
    </tr>
    <tr>
      <th>.dims</th>
      <td><p>An integer. The number of dimensions your are interested in (e.g., 4 for returning the first four principal components).</p></td>
    </tr>
    <tr>
      <th>top</th>
      <td><p>An integer. How many top genes to select for dimensionality reduction</p></td>
    </tr>
    <tr>
      <th>of_samples</th>
      <td><p>A boolean. In case the input is a tidybulk object, it indicates Whether the element column will be sample or transcript column</p></td>
    </tr>
    <tr>
      <th>log_transform</th>
      <td><p>A boolean, whether the value should be log-transformed (e.g., TRUE for RNA sequencing data)</p></td>
    </tr>
    <tr>
      <th>scale</th>
      <td><p>A boolean for method="PCA", this will be passed to the `prcomp` function. It is not included in the ... argument because although the default for `prcomp` if FALSE, it is advisable to set it as TRUE.</p></td>
    </tr>
    <tr>
      <th>action</th>
      <td><p>A character string. Whether to join the new information to the input tbl (add), or just get the non-redundant tbl with the new information (get).</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Further parameters passed to the function prcomp if you choose method="PCA" or Rtsne if you choose method="tSNE"</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A tbl object with additional columns for the reduced dimensions</p>
<p>A tbl object with additional columns for the reduced dimensions</p>
<p>A tbl object with additional columns for the reduced dimensions</p>
<p>A tbl object with additional columns for the reduced dimensions</p>
<p>A `SummarizedExperiment` object</p>
<p>A `SummarizedExperiment` object</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><a href='https://www.tidyverse.org/lifecycle/#maturing'><img src='figures/lifecycle-maturing.svg' alt='Maturing lifecycle'></a></p>
<p>This function reduces the dimensions of the transcript abundances.
It can use multi-dimensional scaling (MDS; DOI.org/10.1186/gb-2010-11-3-r25),
principal component analysis (PCA), or tSNE (Jesse Krijthe et al. 2018)</p>
<p>Underlying method for PCA:
prcomp(scale = scale, ...)</p>
<p>Underlying method for MDS:
limma::plotMDS(ndim = .dims, plot = FALSE, top = top)</p>
<p>Underlying method for tSNE:
Rtsne::Rtsne(data, ...)</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>


<span class='va'>counts.MDS</span> <span class='op'>=</span>  
 <span class='fu'>tidybulk</span><span class='fu'>::</span><span class='va'><a href='counts_mini.html'>counts_mini</a></span> <span class='op'>%&gt;%</span> 
 <span class='fu'><a href='tidybulk.html'>tidybulk</a></span><span class='op'>(</span><span class='va'>sample</span>, <span class='va'>transcript</span>, <span class='va'>count</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
 <span class='fu'><a href='identify_abundant.html'>identify_abundant</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
 <span class='fu'>reduce_dimensions</span><span class='op'>(</span> method<span class='op'>=</span><span class='st'>"MDS"</span>, .dims <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>No group or design set. Assuming all samples belong to one group.</span></div><div class='output co'>#&gt; <span class='message'>tidybulk says: to access the raw results do `attr(..., "internals")$MDS`</span></div><div class='input'>

<span class='va'>counts.PCA</span> <span class='op'>=</span>  
 <span class='fu'>tidybulk</span><span class='fu'>::</span><span class='va'><a href='counts_mini.html'>counts_mini</a></span> <span class='op'>%&gt;%</span> 
 <span class='fu'><a href='tidybulk.html'>tidybulk</a></span><span class='op'>(</span><span class='va'>sample</span>, <span class='va'>transcript</span>, <span class='va'>count</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
 <span class='fu'><a href='identify_abundant.html'>identify_abundant</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
 <span class='fu'>reduce_dimensions</span><span class='op'>(</span>method<span class='op'>=</span><span class='st'>"PCA"</span>, .dims <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>No group or design set. Assuming all samples belong to one group.</span></div><div class='output co'>#&gt; <span class='message'>Getting the 182 most variable genes</span></div><div class='output co'>#&gt; <span class='message'>Fraction of variance explained by the selected principal components</span></div><div class='output co'>#&gt; <span class='message'><span style='color: #555555;'># A tibble: 3 x 2</span><span></span>
#&gt; <span class='message'>  `Fraction of variance`    PC</span>
#&gt; <span class='message'>                   </span><span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span> </span><span style='color: #555555;font-style: italic;'>&lt;int&gt;</span><span></span>
#&gt; <span class='message'></span><span style='color: #555555;'>1</span><span>                 0.598      1</span>
#&gt; <span class='message'></span><span style='color: #555555;'>2</span><span>                 0.302      2</span>
#&gt; <span class='message'></span><span style='color: #555555;'>3</span><span>                 0.085</span><span style='text-decoration: underline;'>2</span><span>     3</span></div><div class='output co'>#&gt; <span class='message'>tidybulk says: to access the raw results do `attr(..., "internals")$PCA`</span></div><div class='input'>


</div></span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


