<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Drop redundant elements (e.g., samples) for which feature (e.g., transcript/gene) abundances are correlated — remove_redundancy • tidybulk</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Drop redundant elements (e.g., samples) for which feature (e.g., transcript/gene) abundances are correlated — remove_redundancy" />
<meta property="og:description" content="remove_redundancy() takes as input a `tbl` formatted as | &amp;lt;SAMPLE&amp;gt; | &amp;lt;TRANSCRIPT&amp;gt; | &amp;lt;COUNT&amp;gt; | &amp;lt;...&amp;gt; | for correlation method or | &amp;lt;DIMENSION 1&amp;gt; | &amp;lt;DIMENSION 2&amp;gt; | &amp;lt;...&amp;gt; | for reduced_dimensions method, and returns a `tbl` with dropped elements (e.g., samples)." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidybulk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/comparison_with_base_R.html">Comparison with base R</a>
    </li>
    <li>
      <a href="../articles/introduction.html">Overview of the tidybulk package</a>
    </li>
    <li>
      <a href="../articles/manuscript_differential_transcript_abundance.html">Manuscript code - differential transcript abundance</a>
    </li>
    <li>
      <a href="../articles/manuscript_transcriptional_signatures.html">Manuscript code - transcriptional signature identification</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Drop redundant elements (e.g., samples) for which feature (e.g., transcript/gene) abundances are correlated</h1>
    
    <div class="hidden name"><code>remove_redundancy-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>remove_redundancy() takes as input a `tbl` formatted as | &lt;SAMPLE&gt; | &lt;TRANSCRIPT&gt; | &lt;COUNT&gt; | &lt;...&gt; | for correlation method or | &lt;DIMENSION 1&gt; | &lt;DIMENSION 2&gt; | &lt;...&gt; | for reduced_dimensions method, and returns a `tbl` with dropped elements (e.g., samples).</p>
    </div>

    <pre class="usage"><span class='fu'>remove_redundancy</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  .element <span class='op'>=</span> <span class='cn'>NULL</span>,
  .feature <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>method</span>,
  of_samples <span class='op'>=</span> <span class='cn'>TRUE</span>,
  correlation_threshold <span class='op'>=</span> <span class='fl'>0.9</span>,
  top <span class='op'>=</span> <span class='cn'>Inf</span>,
  log_transform <span class='op'>=</span> <span class='cn'>FALSE</span>,
  <span class='va'>Dim_a_column</span>,
  <span class='va'>Dim_b_column</span>
<span class='op'>)</span>

<span class='co'># S4 method for spec_tbl_df</span>
<span class='fu'>remove_redundancy</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  .element <span class='op'>=</span> <span class='cn'>NULL</span>,
  .feature <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>method</span>,
  of_samples <span class='op'>=</span> <span class='cn'>TRUE</span>,
  correlation_threshold <span class='op'>=</span> <span class='fl'>0.9</span>,
  top <span class='op'>=</span> <span class='cn'>Inf</span>,
  log_transform <span class='op'>=</span> <span class='cn'>FALSE</span>,
  Dim_a_column <span class='op'>=</span> <span class='cn'>NULL</span>,
  Dim_b_column <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span>

<span class='co'># S4 method for tbl_df</span>
<span class='fu'>remove_redundancy</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  .element <span class='op'>=</span> <span class='cn'>NULL</span>,
  .feature <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>method</span>,
  of_samples <span class='op'>=</span> <span class='cn'>TRUE</span>,
  correlation_threshold <span class='op'>=</span> <span class='fl'>0.9</span>,
  top <span class='op'>=</span> <span class='cn'>Inf</span>,
  log_transform <span class='op'>=</span> <span class='cn'>FALSE</span>,
  Dim_a_column <span class='op'>=</span> <span class='cn'>NULL</span>,
  Dim_b_column <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span>

<span class='co'># S4 method for tidybulk</span>
<span class='fu'>remove_redundancy</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  .element <span class='op'>=</span> <span class='cn'>NULL</span>,
  .feature <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>method</span>,
  of_samples <span class='op'>=</span> <span class='cn'>TRUE</span>,
  correlation_threshold <span class='op'>=</span> <span class='fl'>0.9</span>,
  top <span class='op'>=</span> <span class='cn'>Inf</span>,
  log_transform <span class='op'>=</span> <span class='cn'>FALSE</span>,
  Dim_a_column <span class='op'>=</span> <span class='cn'>NULL</span>,
  Dim_b_column <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span>

<span class='co'># S4 method for SummarizedExperiment</span>
<span class='fu'>remove_redundancy</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  .element <span class='op'>=</span> <span class='cn'>NULL</span>,
  .feature <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>method</span>,
  of_samples <span class='op'>=</span> <span class='cn'>TRUE</span>,
  correlation_threshold <span class='op'>=</span> <span class='fl'>0.9</span>,
  top <span class='op'>=</span> <span class='cn'>Inf</span>,
  log_transform <span class='op'>=</span> <span class='cn'>FALSE</span>,
  Dim_a_column <span class='op'>=</span> <span class='cn'>NULL</span>,
  Dim_b_column <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span>

<span class='co'># S4 method for RangedSummarizedExperiment</span>
<span class='fu'>remove_redundancy</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  .element <span class='op'>=</span> <span class='cn'>NULL</span>,
  .feature <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>method</span>,
  of_samples <span class='op'>=</span> <span class='cn'>TRUE</span>,
  correlation_threshold <span class='op'>=</span> <span class='fl'>0.9</span>,
  top <span class='op'>=</span> <span class='cn'>Inf</span>,
  log_transform <span class='op'>=</span> <span class='cn'>FALSE</span>,
  Dim_a_column <span class='op'>=</span> <span class='cn'>NULL</span>,
  Dim_b_column <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>.data</th>
      <td><p>A `tbl` formatted as | &lt;SAMPLE&gt; | &lt;TRANSCRIPT&gt; | &lt;COUNT&gt; | &lt;...&gt; |</p></td>
    </tr>
    <tr>
      <th>.element</th>
      <td><p>The name of the element column (normally samples).</p></td>
    </tr>
    <tr>
      <th>.feature</th>
      <td><p>The name of the feature column (normally transcripts/genes)</p></td>
    </tr>
    <tr>
      <th>.abundance</th>
      <td><p>The name of the column including the numerical value the clustering is based on (normally transcript abundance)</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>A character string. The cluster algorithm to use, ay the moment k-means is the only algorithm included.</p></td>
    </tr>
    <tr>
      <th>of_samples</th>
      <td><p>A boolean. In case the input is a tidybulk object, it indicates Whether the element column will be sample or transcript column</p></td>
    </tr>
    <tr>
      <th>correlation_threshold</th>
      <td><p>A real number between 0 and 1. For correlation based calculation.</p></td>
    </tr>
    <tr>
      <th>top</th>
      <td><p>An integer. How many top genes to select for correlation based method</p></td>
    </tr>
    <tr>
      <th>log_transform</th>
      <td><p>A boolean, whether the value should be log-transformed (e.g., TRUE for RNA sequencing data)</p></td>
    </tr>
    <tr>
      <th>Dim_a_column</th>
      <td><p>A character string. For reduced_dimension based calculation. The column of one principal component</p></td>
    </tr>
    <tr>
      <th>Dim_b_column</th>
      <td><p>A character string. For reduced_dimension based calculation. The column of another principal component</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A tbl object with with dropped redundant elements (e.g., samples).</p>
<p>A tbl object with with dropped redundant elements (e.g., samples).</p>
<p>A tbl object with with dropped redundant elements (e.g., samples).</p>
<p>A tbl object with with dropped redundant elements (e.g., samples).</p>
<p>A `SummarizedExperiment` object</p>
<p>A `SummarizedExperiment` object</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    



    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>

 <span class='fu'>tidybulk</span><span class='fu'>::</span><span class='va'><a href='counts_mini.html'>counts_mini</a></span> <span class='op'>%&gt;%</span> 
 <span class='fu'><a href='tidybulk.html'>tidybulk</a></span><span class='op'>(</span><span class='va'>sample</span>, <span class='va'>transcript</span>, <span class='va'>count</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
 <span class='fu'><a href='identify_abundant.html'>identify_abundant</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
   <span class='fu'>remove_redundancy</span><span class='op'>(</span>
     .element <span class='op'>=</span> <span class='va'>sample</span>,
     .feature <span class='op'>=</span> <span class='va'>transcript</span>,
       .abundance <span class='op'>=</span>  <span class='va'>count</span>,
       method <span class='op'>=</span> <span class='st'>"correlation"</span>
       <span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>No group or design set. Assuming all samples belong to one group.</span></div><div class='output co'>#&gt; <span class='message'>Getting the 527 most variable genes</span></div><div class='output co'>#&gt; <span class='warning'>Warning: `tbl_df()` is deprecated as of dplyr 1.0.0.</span>
#&gt; <span class='warning'>Please use `tibble::as_tibble()` instead.</span>
#&gt; <span class='warning'><span style='color: #555555;'>This warning is displayed once every 8 hours.</span><span></span>
#&gt; <span class='warning'></span><span style='color: #555555;'>Call `lifecycle::last_warnings()` to see where this warning was generated.</span><span></span></div><div class='output co'>#&gt; </span><span style='color: #555555;'># A tibble: 2,108 x 7</span><span>
#&gt;    sample     transcript `Cell type` count time  condition .abundant
#&gt;    </span><span style='color: #555555;font-style: italic;'>&lt;chr&gt;</span><span>      </span><span style='color: #555555;font-style: italic;'>&lt;chr&gt;</span><span>      </span><span style='color: #555555;font-style: italic;'>&lt;chr&gt;</span><span>       </span><span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span> </span><span style='color: #555555;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #555555;font-style: italic;'>&lt;lgl&gt;</span><span>     </span><span style='color: #555555;font-style: italic;'>&lt;lgl&gt;</span><span>    
#&gt; </span><span style='color: #555555;'> 1</span><span> SRR1740035 TNFRSF4    b_cell          6 1 d   TRUE      FALSE    
#&gt; </span><span style='color: #555555;'> 2</span><span> SRR1740035 PLCH2      b_cell       </span><span style='text-decoration: underline;'>1</span><span>134 1 d   TRUE      TRUE     
#&gt; </span><span style='color: #555555;'> 3</span><span> SRR1740035 PADI4      b_cell         17 1 d   TRUE      TRUE     
#&gt; </span><span style='color: #555555;'> 4</span><span> SRR1740035 PAX7       b_cell          0 1 d   TRUE      FALSE    
#&gt; </span><span style='color: #555555;'> 5</span><span> SRR1740035 CDA        b_cell          0 1 d   TRUE      FALSE    
#&gt; </span><span style='color: #555555;'> 6</span><span> SRR1740035 RCAN3      b_cell        832 1 d   TRUE      TRUE     
#&gt; </span><span style='color: #555555;'> 7</span><span> SRR1740035 SMPDL3B    b_cell          5 1 d   TRUE      FALSE    
#&gt; </span><span style='color: #555555;'> 8</span><span> SRR1740035 EPB41      b_cell       </span><span style='text-decoration: underline;'>4</span><span>709 1 d   TRUE      TRUE     
#&gt; </span><span style='color: #555555;'> 9</span><span> SRR1740035 LCK        b_cell        507 1 d   TRUE      TRUE     
#&gt; </span><span style='color: #555555;'>10</span><span> SRR1740035 COL8A2     b_cell          5 1 d   TRUE      FALSE    
#&gt; </span><span style='color: #555555;'># … with 2,098 more rows</span><span></div><div class='input'>
<span class='va'>counts.MDS</span> <span class='op'>=</span> 
 <span class='fu'>tidybulk</span><span class='fu'>::</span><span class='va'><a href='counts_mini.html'>counts_mini</a></span> <span class='op'>%&gt;%</span> 
 <span class='fu'><a href='tidybulk.html'>tidybulk</a></span><span class='op'>(</span><span class='va'>sample</span>, <span class='va'>transcript</span>, <span class='va'>count</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
 <span class='fu'><a href='identify_abundant.html'>identify_abundant</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>%&gt;%</span> 
  <span class='fu'><a href='reduce_dimensions.html'>reduce_dimensions</a></span><span class='op'>(</span> method<span class='op'>=</span><span class='st'>"MDS"</span>, .dims <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>No group or design set. Assuming all samples belong to one group.</span></div><div class='output co'>#&gt; <span class='message'>tidybulk says: to access the raw results do `attr(..., "internals")$MDS`</span></div><div class='input'>
<span class='fu'>remove_redundancy</span><span class='op'>(</span>
  <span class='va'>counts.MDS</span>,
  Dim_a_column <span class='op'>=</span> <span class='va'>`Dim1`</span>,
  Dim_b_column <span class='op'>=</span> <span class='va'>`Dim2`</span>,
  .element <span class='op'>=</span> <span class='va'>sample</span>,
  method <span class='op'>=</span> <span class='st'>"reduced_dimensions"</span>
<span class='op'>)</span>
</div><div class='output co'>#&gt; </span><span style='color: #555555;'># A tibble: 1,581 x 10</span><span>
#&gt;    sample transcript `Cell type` count time  condition .abundant   Dim1   Dim2
#&gt;    </span><span style='color: #555555;font-style: italic;'>&lt;chr&gt;</span><span>  </span><span style='color: #555555;font-style: italic;'>&lt;chr&gt;</span><span>      </span><span style='color: #555555;font-style: italic;'>&lt;chr&gt;</span><span>       </span><span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span> </span><span style='color: #555555;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #555555;font-style: italic;'>&lt;lgl&gt;</span><span>     </span><span style='color: #555555;font-style: italic;'>&lt;lgl&gt;</span><span>      </span><span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span>  </span><span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span>
#&gt; </span><span style='color: #555555;'> 1</span><span> SRR17… TNFRSF4    b_cell          6 1 d   TRUE      FALSE     -</span><span style='color: #BB0000;'>0.898</span><span> -</span><span style='color: #BB0000;'>0.724</span><span>
#&gt; </span><span style='color: #555555;'> 2</span><span> SRR17… PLCH2      b_cell       </span><span style='text-decoration: underline;'>1</span><span>134 1 d   TRUE      TRUE      -</span><span style='color: #BB0000;'>0.898</span><span> -</span><span style='color: #BB0000;'>0.724</span><span>
#&gt; </span><span style='color: #555555;'> 3</span><span> SRR17… PADI4      b_cell         17 1 d   TRUE      TRUE      -</span><span style='color: #BB0000;'>0.898</span><span> -</span><span style='color: #BB0000;'>0.724</span><span>
#&gt; </span><span style='color: #555555;'> 4</span><span> SRR17… PAX7       b_cell          0 1 d   TRUE      FALSE     -</span><span style='color: #BB0000;'>0.898</span><span> -</span><span style='color: #BB0000;'>0.724</span><span>
#&gt; </span><span style='color: #555555;'> 5</span><span> SRR17… CDA        b_cell          0 1 d   TRUE      FALSE     -</span><span style='color: #BB0000;'>0.898</span><span> -</span><span style='color: #BB0000;'>0.724</span><span>
#&gt; </span><span style='color: #555555;'> 6</span><span> SRR17… RCAN3      b_cell        832 1 d   TRUE      TRUE      -</span><span style='color: #BB0000;'>0.898</span><span> -</span><span style='color: #BB0000;'>0.724</span><span>
#&gt; </span><span style='color: #555555;'> 7</span><span> SRR17… SMPDL3B    b_cell          5 1 d   TRUE      FALSE     -</span><span style='color: #BB0000;'>0.898</span><span> -</span><span style='color: #BB0000;'>0.724</span><span>
#&gt; </span><span style='color: #555555;'> 8</span><span> SRR17… EPB41      b_cell       </span><span style='text-decoration: underline;'>4</span><span>709 1 d   TRUE      TRUE      -</span><span style='color: #BB0000;'>0.898</span><span> -</span><span style='color: #BB0000;'>0.724</span><span>
#&gt; </span><span style='color: #555555;'> 9</span><span> SRR17… LCK        b_cell        507 1 d   TRUE      TRUE      -</span><span style='color: #BB0000;'>0.898</span><span> -</span><span style='color: #BB0000;'>0.724</span><span>
#&gt; </span><span style='color: #555555;'>10</span><span> SRR17… COL8A2     b_cell          5 1 d   TRUE      FALSE     -</span><span style='color: #BB0000;'>0.898</span><span> -</span><span style='color: #BB0000;'>0.724</span><span>
#&gt; </span><span style='color: #555555;'># … with 1,571 more rows, and 1 more variable: Dim3 </span><span style='color: #555555;font-style: italic;'>&lt;dbl&gt;</span><span></div><div class='input'>
</div></span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


