<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Scale the counts of transcripts/genes — scale_abundance • tidybulk</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Scale the counts of transcripts/genes — scale_abundance" />
<meta property="og:description" content="scale_abundance() takes as input A `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment)) and Scales transcript abundance compansating for sequencing depth (e.g., with TMM algorithm, Robinson and Oshlack doi.org/10.1186/gb-2010-11-3-r25)." />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidybulk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.5.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/comparison_with_base_R.html">Comparison with base R</a>
    </li>
    <li>
      <a href="../articles/introduction.html">Overview of the tidybulk package</a>
    </li>
    <li>
      <a href="../articles/manuscript_differential_transcript_abundance.html">Manuscript code - differential feature abundance</a>
    </li>
    <li>
      <a href="../articles/manuscript_transcriptional_signatures.html">Manuscript code - transcriptional signature identification</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stemangiola/tidybulk/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Scale the counts of transcripts/genes</h1>
    <small class="dont-index">Source: <a href='https://github.com/stemangiola/tidybulk/blob/master/R/methods.R'><code>R/methods.R</code></a>, <a href='https://github.com/stemangiola/tidybulk/blob/master/R/methods_SE.R'><code>R/methods_SE.R</code></a></small>
    <div class="hidden name"><code>scale_abundance-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>scale_abundance() takes as input A `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment)) and Scales transcript abundance compansating for sequencing depth (e.g., with TMM algorithm, Robinson and Oshlack doi.org/10.1186/gb-2010-11-3-r25).</p>
    </div>

    <div class="ref-usage sourceCode"><pre class='sourceCode r'><code><span class='fu'>scale_abundance</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  .transcript <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  method <span class='op'>=</span> <span class='st'>"TMM"</span>,
  reference_sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  reference_selection_function <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span>

<span class='co'># S4 method for spec_tbl_df</span>
<span class='fu'>scale_abundance</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  .transcript <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  method <span class='op'>=</span> <span class='st'>"TMM"</span>,
  reference_sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  reference_selection_function <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span>

<span class='co'># S4 method for tbl_df</span>
<span class='fu'>scale_abundance</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  .transcript <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  method <span class='op'>=</span> <span class='st'>"TMM"</span>,
  reference_sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  reference_selection_function <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span>

<span class='co'># S4 method for tidybulk</span>
<span class='fu'>scale_abundance</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  .transcript <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  method <span class='op'>=</span> <span class='st'>"TMM"</span>,
  reference_sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  reference_selection_function <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span>

<span class='co'># S4 method for SummarizedExperiment</span>
<span class='fu'>scale_abundance</span><span class='op'>(</span><span class='va'>.data</span>, method <span class='op'>=</span> <span class='st'>"TMM"</span>, reference_sample <span class='op'>=</span> <span class='cn'>NULL</span><span class='op'>)</span>

<span class='co'># S4 method for RangedSummarizedExperiment</span>
<span class='fu'>scale_abundance</span><span class='op'>(</span><span class='va'>.data</span>, method <span class='op'>=</span> <span class='st'>"TMM"</span>, reference_sample <span class='op'>=</span> <span class='cn'>NULL</span><span class='op'>)</span></code></pre></div>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>.data</th>
      <td><p>A `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment))</p></td>
    </tr>
    <tr>
      <th>.sample</th>
      <td><p>The name of the sample column</p></td>
    </tr>
    <tr>
      <th>.transcript</th>
      <td><p>The name of the transcript/gene column</p></td>
    </tr>
    <tr>
      <th>.abundance</th>
      <td><p>The name of the transcript/gene abundance column</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>A character string. The scaling method passed to the back-end function (i.e., edgeR::calcNormFactors; "TMM","TMMwsp","RLE","upperquartile")</p></td>
    </tr>
    <tr>
      <th>reference_sample</th>
      <td><p>A character string. The name of the reference sample. If NULL the sample with highest total read count will be selected as reference.</p></td>
    </tr>
    <tr>
      <th>action</th>
      <td><p>A character string between "add" (default) and "only". "add" joins the new information to the input tbl (default), "only" return a non-redundant tbl with the just new information.</p></td>
    </tr>
    <tr>
      <th>reference_selection_function</th>
      <td><p>DEPRECATED. please use reference_sample.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A tbl object with additional columns with scaled data as `&lt;NAME OF COUNT COLUMN&gt;_scaled`</p>
<p>A tbl object with additional columns with scaled data as `&lt;NAME OF COUNT COLUMN&gt;_scaled`</p>
<p>A tbl object with additional columns with scaled data as `&lt;NAME OF COUNT COLUMN&gt;_scaled`</p>
<p>A tbl object with additional columns with scaled data as `&lt;NAME OF COUNT COLUMN&gt;_scaled`</p>
<p>A `SummarizedExperiment` object</p>
<p>A `SummarizedExperiment` object</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>`r lifecycle::badge("maturing")`</p>
<p>Scales transcript abundance compensating for sequencing depth
(e.g., with TMM algorithm, Robinson and Oshlack doi.org/10.1186/gb-2010-11-3-r25).
Lowly transcribed transcripts/genes (defined with minimum_counts and minimum_proportion parameters)
are filtered out from the scaling procedure.
The scaling inference is then applied back to all unfiltered data.</p>
<p>Underlying method
edgeR::calcNormFactors(.data, method = c("TMM","TMMwsp","RLE","upperquartile"))</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <div class="ref-examples sourceCode"><pre class='sourceCode r'><code><span class='r-in'></span>
<span class='r-in'></span>
<span class='r-in'> <span class='fu'>tidybulk</span><span class='fu'>::</span><span class='va'><a href='se_mini.html'>se_mini</a></span> <span class='op'><a href='pipe.html'>%&gt;%</a></span></span>
<span class='r-in'>   <span class='fu'><a href='identify_abundant-methods.html'>identify_abundant</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'><a href='pipe.html'>%&gt;%</a></span></span>
<span class='r-in'>   <span class='fu'>scale_abundance</span><span class='op'>(</span><span class='op'>)</span></span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> No group or design set. Assuming all samples belong to one group.</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> class: SummarizedExperiment </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> dim: 527 5 </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> metadata(0):</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> assays(2): count count_scaled</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> rownames(527): ABCB4 ABCB9 ... ZNF324 ZNF442</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> rowData names(1): .abundant</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> colnames(5): SRR1740034 SRR1740035 SRR1740043 SRR1740058 SRR1740067</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> colData names(5): Cell.type time condition TMM multiplier</span>
<span class='r-in'></span>
<span class='r-in'></span>
<span class='r-in'></span>
</code></pre></div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by Stefano Mangiola.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


