<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Perform differential transcription testing using edgeR quasi-likelihood (QLT), edgeR likelihood-ratio (LR), limma-voom, limma-voom-with-quality-weights or DESeq2 — test_differential_abundance • tidybulk</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Perform differential transcription testing using edgeR quasi-likelihood (QLT), edgeR likelihood-ratio (LR), limma-voom, limma-voom-with-quality-weights or DESeq2 — test_differential_abundance" />
<meta property="og:description" content="test_differential_abundance() takes as input a `tbl` formatted as | &amp;lt;SAMPLE&amp;gt; | &amp;lt;TRANSCRIPT&amp;gt; | &amp;lt;COUNT&amp;gt; | &amp;lt;...&amp;gt; | and returns a `tbl` with additional columns for the statistics from the hypothesis test." />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



<!-- dependencies from examples -->


  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidybulk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/comparison_with_base_R.html">Comparison with base R</a>
    </li>
    <li>
      <a href="../articles/introduction.html">Overview of the tidybulk package</a>
    </li>
    <li>
      <a href="../articles/manuscript_differential_transcript_abundance.html">Manuscript code - differential feature abundance</a>
    </li>
    <li>
      <a href="../articles/manuscript_transcriptional_signatures.html">Manuscript code - transcriptional signature identification</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform differential transcription testing using edgeR quasi-likelihood (QLT), edgeR likelihood-ratio (LR), limma-voom, limma-voom-with-quality-weights or DESeq2</h1>
    
    <div class="hidden name"><code>test_differential_abundance-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>test_differential_abundance() takes as input a `tbl` formatted as | &lt;SAMPLE&gt; | &lt;TRANSCRIPT&gt; | &lt;COUNT&gt; | &lt;...&gt; | and returns a `tbl` with additional columns for the statistics from the hypothesis test.</p>
    </div>

    <pre class="usage"><span class='fu'><a href='test_differential_abundance-methods.html'>test_differential_abundance</a></span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.formula</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  .transcript <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  .contrasts <span class='op'>=</span> <span class='cn'>NULL</span>,
  method <span class='op'>=</span> <span class='st'>"edgeR_quasi_likelihood"</span>,
  test_above_log2_fold_change <span class='op'>=</span> <span class='cn'>NULL</span>,
  scaling_method <span class='op'>=</span> <span class='st'>"TMM"</span>,
  omit_contrast_in_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
  prefix <span class='op'>=</span> <span class='st'>""</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  <span class='va'>...</span>,
  significance_threshold <span class='op'>=</span> <span class='cn'>NULL</span>,
  fill_missing_values <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span>

<span class='co'># S4 method for spec_tbl_df</span>
<span class='fu'><a href='test_differential_abundance-methods.html'>test_differential_abundance</a></span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.formula</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  .transcript <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  .contrasts <span class='op'>=</span> <span class='cn'>NULL</span>,
  method <span class='op'>=</span> <span class='st'>"edgeR_quasi_likelihood"</span>,
  test_above_log2_fold_change <span class='op'>=</span> <span class='cn'>NULL</span>,
  scaling_method <span class='op'>=</span> <span class='st'>"TMM"</span>,
  omit_contrast_in_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
  prefix <span class='op'>=</span> <span class='st'>""</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  <span class='va'>...</span>,
  significance_threshold <span class='op'>=</span> <span class='cn'>NULL</span>,
  fill_missing_values <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span>

<span class='co'># S4 method for tbl_df</span>
<span class='fu'><a href='test_differential_abundance-methods.html'>test_differential_abundance</a></span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.formula</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  .transcript <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  .contrasts <span class='op'>=</span> <span class='cn'>NULL</span>,
  method <span class='op'>=</span> <span class='st'>"edgeR_quasi_likelihood"</span>,
  test_above_log2_fold_change <span class='op'>=</span> <span class='cn'>NULL</span>,
  scaling_method <span class='op'>=</span> <span class='st'>"TMM"</span>,
  omit_contrast_in_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
  prefix <span class='op'>=</span> <span class='st'>""</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  <span class='va'>...</span>,
  significance_threshold <span class='op'>=</span> <span class='cn'>NULL</span>,
  fill_missing_values <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span>

<span class='co'># S4 method for tidybulk</span>
<span class='fu'><a href='test_differential_abundance-methods.html'>test_differential_abundance</a></span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.formula</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  .transcript <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  .contrasts <span class='op'>=</span> <span class='cn'>NULL</span>,
  method <span class='op'>=</span> <span class='st'>"edgeR_quasi_likelihood"</span>,
  test_above_log2_fold_change <span class='op'>=</span> <span class='cn'>NULL</span>,
  scaling_method <span class='op'>=</span> <span class='st'>"TMM"</span>,
  omit_contrast_in_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
  prefix <span class='op'>=</span> <span class='st'>""</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  <span class='va'>...</span>,
  significance_threshold <span class='op'>=</span> <span class='cn'>NULL</span>,
  fill_missing_values <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span>

<span class='co'># S4 method for SummarizedExperiment</span>
<span class='fu'><a href='test_differential_abundance-methods.html'>test_differential_abundance</a></span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.formula</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  .transcript <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  .contrasts <span class='op'>=</span> <span class='cn'>NULL</span>,
  method <span class='op'>=</span> <span class='st'>"edgeR_quasi_likelihood"</span>,
  test_above_log2_fold_change <span class='op'>=</span> <span class='cn'>NULL</span>,
  scaling_method <span class='op'>=</span> <span class='st'>"TMM"</span>,
  omit_contrast_in_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
  prefix <span class='op'>=</span> <span class='st'>""</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  <span class='va'>...</span>,
  significance_threshold <span class='op'>=</span> <span class='cn'>NULL</span>,
  fill_missing_values <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span>

<span class='co'># S4 method for RangedSummarizedExperiment</span>
<span class='fu'><a href='test_differential_abundance-methods.html'>test_differential_abundance</a></span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.formula</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  .transcript <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  .contrasts <span class='op'>=</span> <span class='cn'>NULL</span>,
  method <span class='op'>=</span> <span class='st'>"edgeR_quasi_likelihood"</span>,
  test_above_log2_fold_change <span class='op'>=</span> <span class='cn'>NULL</span>,
  scaling_method <span class='op'>=</span> <span class='st'>"TMM"</span>,
  omit_contrast_in_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
  prefix <span class='op'>=</span> <span class='st'>""</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  <span class='va'>...</span>,
  significance_threshold <span class='op'>=</span> <span class='cn'>NULL</span>,
  fill_missing_values <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>.data</th>
      <td><p>A `tbl` formatted as | &lt;SAMPLE&gt; | &lt;TRANSCRIPT&gt; | &lt;COUNT&gt; | &lt;...&gt; |</p></td>
    </tr>
    <tr>
      <th>.formula</th>
      <td><p>A formula representing the desired linear model. If there is more than one factor, they should be in the order factor of interest + additional factors.</p></td>
    </tr>
    <tr>
      <th>.sample</th>
      <td><p>The name of the sample column</p></td>
    </tr>
    <tr>
      <th>.transcript</th>
      <td><p>The name of the transcript/gene column</p></td>
    </tr>
    <tr>
      <th>.abundance</th>
      <td><p>The name of the transcript/gene abundance column</p></td>
    </tr>
    <tr>
      <th>.contrasts</th>
      <td><p>This parameter takes the format of the contrast parameter of the method of choice. For edgeR and limma-voom is a character vector. For DESeq2 is a list including a character vector of length three. The first covariate is the one the model is tested against (e.g., ~ factor_of_interest)</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>A string character. Either "edgeR_quasi_likelihood" (i.e., QLF), "edgeR_likelihood_ratio" (i.e., LRT), "edger_robust_likelihood_ratio", "DESeq2", "limma_voom", "limma_voom_sample_weights"</p></td>
    </tr>
    <tr>
      <th>test_above_log2_fold_change</th>
      <td><p>A positive real value. This works for edgeR and limma_voom methods. It uses the `treat` function, which tests that the difference in abundance is bigger than this threshold rather than zero <a href='https://pubmed.ncbi.nlm.nih.gov/19176553'>https://pubmed.ncbi.nlm.nih.gov/19176553</a>.</p></td>
    </tr>
    <tr>
      <th>scaling_method</th>
      <td><p>A character string. The scaling method passed to the back-end functions: edgeR and limma-voom (i.e., edgeR::calcNormFactors; "TMM","TMMwsp","RLE","upperquartile"). Setting the parameter to \"none\" will skip the compensation for sequencing-depth for the method edgeR or limma-voom.</p></td>
    </tr>
    <tr>
      <th>omit_contrast_in_colnames</th>
      <td><p>If just one contrast is specified you can choose to omit the contrast label in the colnames.</p></td>
    </tr>
    <tr>
      <th>prefix</th>
      <td><p>A character string. The prefix you would like to add to the result columns. It is useful if you want to compare several methods.</p></td>
    </tr>
    <tr>
      <th>action</th>
      <td><p>A character string. Whether to join the new information to the input tbl (add), or just get the non-redundant tbl with the new information (get).</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Further arguments passed to some of the internal functions. Currently, it is needed just for internal debug.</p></td>
    </tr>
    <tr>
      <th>significance_threshold</th>
      <td><p>DEPRECATED - A real between 0 and 1 (usually 0.05).</p></td>
    </tr>
    <tr>
      <th>fill_missing_values</th>
      <td><p>DEPRECATED - A boolean. Whether to fill missing sample/transcript values with the median of the transcript. This is rarely needed.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A `tbl` with additional columns for the statistics from the test (e.g.,  log fold change, p-value and false discovery rate).</p>
<p>A `tbl` with additional columns for the statistics from the test (e.g.,  log fold change, p-value and false discovery rate).</p>
<p>A `tbl` with additional columns for the statistics from the hypothesis test (e.g.,  log fold change, p-value and false discovery rate).</p>
<p>A `tbl` with additional columns for the statistics from the hypothesis test (e.g.,  log fold change, p-value and false discovery rate).</p>
<p>A `SummarizedExperiment` object</p>
<p>A `SummarizedExperiment` object</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>`r lifecycle::badge("maturing")`</p>
<p>This function provides the option to use edgeR <a href='https://doi.org/10.1093/bioinformatics/btp616'>https://doi.org/10.1093/bioinformatics/btp616</a>, limma-voom <a href='https://doi.org/10.1186/gb-2014-15-2-r29'>https://doi.org/10.1186/gb-2014-15-2-r29</a>, limma_voom_sample_weights <a href='https://doi.org/10.1093/nar/gkv412'>https://doi.org/10.1093/nar/gkv412</a> or  DESeq2 <a href='https://doi.org/10.1186/s13059-014-0550-8'>https://doi.org/10.1186/s13059-014-0550-8</a> to perform the testing.
All methods use raw counts, irrespective of if scale_abundance or adjust_abundance have been calculated, therefore it is essential to add covariates such as batch effects (if applicable) in the formula.</p>
<p>Underlying method for edgeR framework:
	.data <!-- %&gt;% --></p>
<p># Filter
keep_abundant(
		factor_of_interest = !!(as.symbol(parse_formula(.formula)[1])),
		minimum_counts = minimum_counts,
		minimum_proportion = minimum_proportion
	) <!-- %&gt;% --></p>
<p># Format
		select(!!.transcript,!!.sample,!!.abundance) <!-- %&gt;% -->
		spread(!!.sample,!!.abundance) <!-- %&gt;% -->
		as_matrix(rownames = !!.transcript) <!-- %&gt;% --></p>
<p># edgeR
		edgeR::DGEList(counts = .) <!-- %&gt;% -->
		edgeR::calcNormFactors(method = scaling_method) <!-- %&gt;% -->
		edgeR::estimateDisp(design) <!-- %&gt;% --></p>
<p># Fit
		edgeR::glmQLFit(design) <!-- %&gt;% // or glmFit according to choice -->
		edgeR::glmQLFTest(coef = 2, contrast = my_contrasts) // or glmLRT according to choice</p>
<p>Underlying method for DESeq2 framework:
keep_abundant(
		factor_of_interest = !!as.symbol(parse_formula(.formula)[[1]]),
		minimum_counts = minimum_counts,
		minimum_proportion = minimum_proportion
) <!-- %&gt;% --></p>
<p># DESeq2
DESeq2::DESeqDataSet( design = .formula) <!-- %&gt;% -->
DESeq2::DESeq() <!-- %&gt;% -->
DESeq2::results()</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
 <span class='co'># edgeR</span>

 <span class='fu'>tidybulk</span><span class='fu'>::</span><span class='va'><a href='se_mini.html'>se_mini</a></span> <span class='op'>%&gt;%</span>
 <span class='fu'><a href='identify_abundant-methods.html'>identify_abundant</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='test_differential_abundance-methods.html'>test_differential_abundance</a></span><span class='op'>(</span> <span class='op'>~</span> <span class='va'>condition</span> <span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>No group or design set. Assuming all samples belong to one group.</span></div><div class='output co'>#&gt; <span class='message'>=====================================</span>
#&gt; <span class='message'>tidybulk says: All testing methods use raw counts, irrespective of if scale_abundance </span>
#&gt; <span class='message'>or adjust_abundance have been calculated. Therefore, it is essential to add covariates </span>
#&gt; <span class='message'>such as batch effects (if applicable) in the formula.</span>
#&gt; <span class='message'>=====================================</span></div><div class='output co'>#&gt; <span class='message'>tidybulk says: The design column names are "(Intercept), conditionTRUE"</span></div><div class='output co'>#&gt; <span class='message'>tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$edgeR_quasi_likelihood`</span></div><div class='output co'>#&gt; class: SummarizedExperiment 
#&gt; dim: 527 5 
#&gt; metadata(0):
#&gt; assays(1): count
#&gt; rownames(527): ABCB4 ABCB9 ... ZNF324 ZNF442
#&gt; rowData names(6): .abundant logFC ... PValue FDR
#&gt; colnames(5): SRR1740034 SRR1740035 SRR1740043 SRR1740058 SRR1740067
#&gt; colData names(3): Cell.type time condition</div><div class='input'>
  <span class='co'># The function `test_differential_abundance` operates with contrasts too</span>

 <span class='fu'>tidybulk</span><span class='fu'>::</span><span class='va'><a href='se_mini.html'>se_mini</a></span> <span class='op'>%&gt;%</span>
 <span class='fu'><a href='identify_abundant-methods.html'>identify_abundant</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
 <span class='fu'><a href='test_differential_abundance-methods.html'>test_differential_abundance</a></span><span class='op'>(</span>
      <span class='op'>~</span> <span class='fl'>0</span> <span class='op'>+</span> <span class='va'>condition</span>,
      .contrasts <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span> <span class='st'>"conditionTRUE - conditionFALSE"</span><span class='op'>)</span>
 <span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>No group or design set. Assuming all samples belong to one group.</span></div><div class='output co'>#&gt; <span class='message'>=====================================</span>
#&gt; <span class='message'>tidybulk says: All testing methods use raw counts, irrespective of if scale_abundance </span>
#&gt; <span class='message'>or adjust_abundance have been calculated. Therefore, it is essential to add covariates </span>
#&gt; <span class='message'>such as batch effects (if applicable) in the formula.</span>
#&gt; <span class='message'>=====================================</span></div><div class='output co'>#&gt; <span class='message'>tidybulk says: The design column names are "conditionFALSE, conditionTRUE"</span></div><div class='output co'>#&gt; <span class='message'>tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$edgeR_quasi_likelihood`</span></div><div class='output co'>#&gt; class: SummarizedExperiment 
#&gt; dim: 527 5 
#&gt; metadata(0):
#&gt; assays(1): count
#&gt; rownames(527): ABCB4 ABCB9 ... ZNF324 ZNF442
#&gt; rowData names(6): .abundant logFC___conditionTRUE - conditionFALSE ...
#&gt;   PValue___conditionTRUE - conditionFALSE FDR___conditionTRUE -
#&gt;   conditionFALSE
#&gt; colnames(5): SRR1740034 SRR1740035 SRR1740043 SRR1740058 SRR1740067
#&gt; colData names(3): Cell.type time condition</div><div class='input'>
 <span class='co'># DESeq2 - equivalent for limma-voom</span>

<span class='va'>my_se_mini</span> <span class='op'>=</span> <span class='fu'>tidybulk</span><span class='fu'>::</span><span class='va'><a href='se_mini.html'>se_mini</a></span>
<span class='va'>my_se_mini</span><span class='op'>$</span><span class='va'>condition</span>  <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='va'>my_se_mini</span><span class='op'>$</span><span class='va'>condition</span><span class='op'>)</span>

<span class='va'>my_se_mini</span>  <span class='op'>%&gt;%</span>
 <span class='fu'><a href='identify_abundant-methods.html'>identify_abundant</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='test_differential_abundance-methods.html'>test_differential_abundance</a></span><span class='op'>(</span> <span class='op'>~</span> <span class='va'>condition</span>, method<span class='op'>=</span><span class='st'>"deseq2"</span> <span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>No group or design set. Assuming all samples belong to one group.</span></div><div class='output co'>#&gt; <span class='message'>=====================================</span>
#&gt; <span class='message'>tidybulk says: All testing methods use raw counts, irrespective of if scale_abundance </span>
#&gt; <span class='message'>or adjust_abundance have been calculated. Therefore, it is essential to add covariates </span>
#&gt; <span class='message'>such as batch effects (if applicable) in the formula.</span>
#&gt; <span class='message'>=====================================</span></div><div class='output co'>#&gt; <span class='message'>renaming the first element in assays to 'counts'</span></div><div class='output co'>#&gt; <span class='message'>converting counts to integer mode</span></div><div class='output co'>#&gt; <span class='message'>estimating size factors</span></div><div class='output co'>#&gt; <span class='message'>estimating dispersions</span></div><div class='output co'>#&gt; <span class='message'>gene-wise dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>mean-dispersion relationship</span></div><div class='output co'>#&gt; <span class='message'>-- note: fitType='parametric', but the dispersion trend was not well captured by the</span>
#&gt; <span class='message'>   function: y = a/x + b, and a local regression fit was automatically substituted.</span>
#&gt; <span class='message'>   specify fitType='local' or 'mean' to avoid this message next time.</span></div><div class='output co'>#&gt; <span class='message'>final dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>fitting model and testing</span></div><div class='output co'>#&gt; <span class='message'>tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$deseq2`</span></div><div class='output co'>#&gt; class: SummarizedExperiment 
#&gt; dim: 527 5 
#&gt; metadata(0):
#&gt; assays(1): count
#&gt; rownames(527): ABCB4 ABCB9 ... ZNF324 ZNF442
#&gt; rowData names(7): .abundant baseMean ... pvalue padj
#&gt; colnames(5): SRR1740034 SRR1740035 SRR1740043 SRR1740058 SRR1740067
#&gt; colData names(3): Cell.type time condition</div><div class='input'>
  <span class='co'># The function `test_differential_abundance` operates with contrasts too</span>

 <span class='va'>my_se_mini</span> <span class='op'>%&gt;%</span>
 <span class='fu'><a href='identify_abundant-methods.html'>identify_abundant</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
 <span class='fu'><a href='test_differential_abundance-methods.html'>test_differential_abundance</a></span><span class='op'>(</span>
      <span class='op'>~</span> <span class='fl'>0</span> <span class='op'>+</span> <span class='va'>condition</span>,
      .contrasts <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"condition"</span>, <span class='st'>"TRUE"</span>, <span class='st'>"FALSE"</span><span class='op'>)</span><span class='op'>)</span>,
      method<span class='op'>=</span><span class='st'>"deseq2"</span>
 <span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>No group or design set. Assuming all samples belong to one group.</span></div><div class='output co'>#&gt; <span class='message'>=====================================</span>
#&gt; <span class='message'>tidybulk says: All testing methods use raw counts, irrespective of if scale_abundance </span>
#&gt; <span class='message'>or adjust_abundance have been calculated. Therefore, it is essential to add covariates </span>
#&gt; <span class='message'>such as batch effects (if applicable) in the formula.</span>
#&gt; <span class='message'>=====================================</span></div><div class='output co'>#&gt; <span class='message'>renaming the first element in assays to 'counts'</span></div><div class='output co'>#&gt; <span class='message'>converting counts to integer mode</span></div><div class='output co'>#&gt; <span class='message'>estimating size factors</span></div><div class='output co'>#&gt; <span class='message'>estimating dispersions</span></div><div class='output co'>#&gt; <span class='message'>gene-wise dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>mean-dispersion relationship</span></div><div class='output co'>#&gt; <span class='message'>-- note: fitType='parametric', but the dispersion trend was not well captured by the</span>
#&gt; <span class='message'>   function: y = a/x + b, and a local regression fit was automatically substituted.</span>
#&gt; <span class='message'>   specify fitType='local' or 'mean' to avoid this message next time.</span></div><div class='output co'>#&gt; <span class='message'>final dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>fitting model and testing</span></div><div class='output co'>#&gt; <span class='message'>tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$deseq2`</span></div><div class='output co'>#&gt; class: SummarizedExperiment 
#&gt; dim: 527 5 
#&gt; metadata(0):
#&gt; assays(1): count
#&gt; rownames(527): ABCB4 ABCB9 ... ZNF324 ZNF442
#&gt; rowData names(7): .abundant baseMean___condition TRUE-FALSE ...
#&gt;   pvalue___condition TRUE-FALSE padj___condition TRUE-FALSE
#&gt; colnames(5): SRR1740034 SRR1740035 SRR1740043 SRR1740058 SRR1740067
#&gt; colData names(3): Cell.type time condition</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by Stefano Mangiola.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


