<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Perform differential transcription testing using edgeR quasi-likelihood (QLT), edgeR likelihood-ratio (LR), limma-voom, limma-voom-with-quality-weights or DESeq2 — test_differential_abundance • tidybulk</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Perform differential transcription testing using edgeR quasi-likelihood (QLT), edgeR likelihood-ratio (LR), limma-voom, limma-voom-with-quality-weights or DESeq2 — test_differential_abundance" />
<meta property="og:description" content="test_differential_abundance() takes as input a `tbl` formatted as | &amp;lt;SAMPLE&amp;gt; | &amp;lt;TRANSCRIPT&amp;gt; | &amp;lt;COUNT&amp;gt; | &amp;lt;...&amp;gt; | and returns a `tbl` with additional columns for the statistics from the hypothesis test." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidybulk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/comparison_with_base_R.html">Comparison with base R</a>
    </li>
    <li>
      <a href="../articles/introduction.html">Overview of the tidybulk package</a>
    </li>
    <li>
      <a href="../articles/manuscript_differential_transcript_abundance.html">Manuscript code - differential transcript abundance</a>
    </li>
    <li>
      <a href="../articles/manuscript_transcriptional_signatures.html">Manuscript code - transcriptional signature identification</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform differential transcription testing using edgeR quasi-likelihood (QLT), edgeR likelihood-ratio (LR), limma-voom, limma-voom-with-quality-weights or DESeq2</h1>
    
    <div class="hidden name"><code>test_differential_abundance-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>test_differential_abundance() takes as input a `tbl` formatted as | &lt;SAMPLE&gt; | &lt;TRANSCRIPT&gt; | &lt;COUNT&gt; | &lt;...&gt; | and returns a `tbl` with additional columns for the statistics from the hypothesis test.</p>
    </div>

    <pre class="usage"><span class='fu'>test_differential_abundance</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.formula</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  .transcript <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  .contrasts <span class='op'>=</span> <span class='cn'>NULL</span>,
  method <span class='op'>=</span> <span class='st'>"edgeR_quasi_likelihood"</span>,
  test_above_log2_fold_change <span class='op'>=</span> <span class='cn'>NULL</span>,
  scaling_method <span class='op'>=</span> <span class='st'>"TMM"</span>,
  omit_contrast_in_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
  prefix <span class='op'>=</span> <span class='st'>""</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  significance_threshold <span class='op'>=</span> <span class='cn'>NULL</span>,
  fill_missing_values <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span>

<span class='co'># S4 method for spec_tbl_df</span>
<span class='fu'>test_differential_abundance</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.formula</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  .transcript <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  .contrasts <span class='op'>=</span> <span class='cn'>NULL</span>,
  method <span class='op'>=</span> <span class='st'>"edgeR_quasi_likelihood"</span>,
  test_above_log2_fold_change <span class='op'>=</span> <span class='cn'>NULL</span>,
  scaling_method <span class='op'>=</span> <span class='st'>"TMM"</span>,
  omit_contrast_in_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
  prefix <span class='op'>=</span> <span class='st'>""</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  significance_threshold <span class='op'>=</span> <span class='cn'>NULL</span>,
  fill_missing_values <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span>

<span class='co'># S4 method for tbl_df</span>
<span class='fu'>test_differential_abundance</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.formula</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  .transcript <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  .contrasts <span class='op'>=</span> <span class='cn'>NULL</span>,
  method <span class='op'>=</span> <span class='st'>"edgeR_quasi_likelihood"</span>,
  test_above_log2_fold_change <span class='op'>=</span> <span class='cn'>NULL</span>,
  scaling_method <span class='op'>=</span> <span class='st'>"TMM"</span>,
  omit_contrast_in_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
  prefix <span class='op'>=</span> <span class='st'>""</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  significance_threshold <span class='op'>=</span> <span class='cn'>NULL</span>,
  fill_missing_values <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span>

<span class='co'># S4 method for tidybulk</span>
<span class='fu'>test_differential_abundance</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.formula</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  .transcript <span class='op'>=</span> <span class='cn'>NULL</span>,
  .abundance <span class='op'>=</span> <span class='cn'>NULL</span>,
  .contrasts <span class='op'>=</span> <span class='cn'>NULL</span>,
  method <span class='op'>=</span> <span class='st'>"edgeR_quasi_likelihood"</span>,
  test_above_log2_fold_change <span class='op'>=</span> <span class='cn'>NULL</span>,
  scaling_method <span class='op'>=</span> <span class='st'>"TMM"</span>,
  omit_contrast_in_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
  prefix <span class='op'>=</span> <span class='st'>""</span>,
  action <span class='op'>=</span> <span class='st'>"add"</span>,
  significance_threshold <span class='op'>=</span> <span class='cn'>NULL</span>,
  fill_missing_values <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span>

<span class='co'># S4 method for SummarizedExperiment</span>
<span class='fu'>test_differential_abundance</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.formula</span>,
  .contrasts <span class='op'>=</span> <span class='cn'>NULL</span>,
  method <span class='op'>=</span> <span class='st'>"edgeR_quasi_likelihood"</span>,
  test_above_log2_fold_change <span class='op'>=</span> <span class='cn'>NULL</span>,
  scaling_method <span class='op'>=</span> <span class='st'>"TMM"</span>,
  omit_contrast_in_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
  prefix <span class='op'>=</span> <span class='st'>""</span>
<span class='op'>)</span>

<span class='co'># S4 method for RangedSummarizedExperiment</span>
<span class='fu'>test_differential_abundance</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.formula</span>,
  .contrasts <span class='op'>=</span> <span class='cn'>NULL</span>,
  method <span class='op'>=</span> <span class='st'>"edgeR_quasi_likelihood"</span>,
  test_above_log2_fold_change <span class='op'>=</span> <span class='cn'>NULL</span>,
  scaling_method <span class='op'>=</span> <span class='st'>"TMM"</span>,
  omit_contrast_in_colnames <span class='op'>=</span> <span class='cn'>FALSE</span>,
  prefix <span class='op'>=</span> <span class='st'>""</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>.data</th>
      <td><p>A `tbl` formatted as | &lt;SAMPLE&gt; | &lt;TRANSCRIPT&gt; | &lt;COUNT&gt; | &lt;...&gt; |</p></td>
    </tr>
    <tr>
      <th>.formula</th>
      <td><p>A formula representing the desired linear model. If there is more than one factor, they should be in the order factor of interest + additional factors.</p></td>
    </tr>
    <tr>
      <th>.sample</th>
      <td><p>The name of the sample column</p></td>
    </tr>
    <tr>
      <th>.transcript</th>
      <td><p>The name of the transcript/gene column</p></td>
    </tr>
    <tr>
      <th>.abundance</th>
      <td><p>The name of the transcript/gene abundance column</p></td>
    </tr>
    <tr>
      <th>.contrasts</th>
      <td><p>This parameter takes the format of the contrast parameter of the method of choice. For edgeR and limma-voom is a character vector. For DESeq2 is a list including a character vector of length three. The first covariate is the one the model is tested against (e.g., ~ factor_of_interest)</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>A string character. Either "edgeR_quasi_likelihood" (i.e., QLF), "edgeR_likelihood_ratio" (i.e., LRT), "edger_robust_likelihood_ratio", "DESeq2", "limma_voom", "limma_voom_sample_weights"</p></td>
    </tr>
    <tr>
      <th>test_above_log2_fold_change</th>
      <td><p>A positive real value. At the moment this works just for edgeR methods, and use the `treat` function, which test the that the difference in abundance is bigger than this parameter rather than zero <a href='https://www.rdocumentation.org/packages/edgeR/versions/3.14.0/topics/glmTreat'>https://www.rdocumentation.org/packages/edgeR/versions/3.14.0/topics/glmTreat</a>.</p></td>
    </tr>
    <tr>
      <th>scaling_method</th>
      <td><p>A character string. The scaling method passed to the back-end functions: edgeR and limma-voom (i.e., edgeR::calcNormFactors; "TMM","TMMwsp","RLE","upperquartile"). Setting the parameter to \"none\" will skip the compensation for sequencing-depth for the method edgeR or limma-voom.</p></td>
    </tr>
    <tr>
      <th>omit_contrast_in_colnames</th>
      <td><p>If just one contrast is specified you can choose to omit the contrast label in the colnames.</p></td>
    </tr>
    <tr>
      <th>prefix</th>
      <td><p>A character string. The prefix you would like to add to the result columns. It is useful if you want to compare several methods.</p></td>
    </tr>
    <tr>
      <th>action</th>
      <td><p>A character string. Whether to join the new information to the input tbl (add), or just get the non-redundant tbl with the new information (get).</p></td>
    </tr>
    <tr>
      <th>significance_threshold</th>
      <td><p>DEPRECATED - A real between 0 and 1 (usually 0.05).</p></td>
    </tr>
    <tr>
      <th>fill_missing_values</th>
      <td><p>DEPRECATED - A boolean. Whether to fill missing sample/transcript values with the median of the transcript. This is rarely needed.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A `tbl` with additional columns for the statistics from the test (e.g.,  log fold change, p-value and false discovery rate).</p>
<p>A `tbl` with additional columns for the statistics from the test (e.g.,  log fold change, p-value and false discovery rate).</p>
<p>A `tbl` with additional columns for the statistics from the hypothesis test (e.g.,  log fold change, p-value and false discovery rate).</p>
<p>A `tbl` with additional columns for the statistics from the hypothesis test (e.g.,  log fold change, p-value and false discovery rate).</p>
<p>A `SummarizedExperiment` object</p>
<p>A `SummarizedExperiment` object</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><a href='https://lifecycle.r-lib.org/articles/stages.html#maturing'><img src='figures/lifecycle-maturing.svg' alt='[Maturing]' /></a></p>
<p>This function provides the option to use edgeR <a href='https://doi.org/10.1093/bioinformatics/btp616'>https://doi.org/10.1093/bioinformatics/btp616</a>, limma-voom <a href='https://doi.org/10.1186/gb-2014-15-2-r29'>https://doi.org/10.1186/gb-2014-15-2-r29</a>, limma_voom_sample_weights <a href='https://doi.org/10.1093/nar/gkv412'>https://doi.org/10.1093/nar/gkv412</a> or  DESeq2 <a href='https://doi.org/10.1186/s13059-014-0550-8'>https://doi.org/10.1186/s13059-014-0550-8</a> to perform the testing.
All methods use raw counts, irrespective of if scale_abundance or adjust_abundance have been calculated, therefore it is essential to add covariates such as batch effects (if applicable) in the formula.</p>
<p>Underlying method for edgeR framework:
	.data <!-- %&gt;% --></p>
<p># Filter
keep_abundant(
		factor_of_interest = !!(as.symbol(parse_formula(.formula)[1])),
		minimum_counts = minimum_counts,
		minimum_proportion = minimum_proportion
	) <!-- %&gt;% --></p>
<p># Format
		select(!!.transcript,!!.sample,!!.abundance) <!-- %&gt;% -->
		spread(!!.sample,!!.abundance) <!-- %&gt;% -->
		as_matrix(rownames = !!.transcript) <!-- %&gt;% --></p>
<p># edgeR
		edgeR::DGEList(counts = .) <!-- %&gt;% -->
		edgeR::calcNormFactors(method = scaling_method) <!-- %&gt;% -->
		edgeR::estimateDisp(design) <!-- %&gt;% --></p>
<p># Fit
		edgeR::glmQLFit(design) <!-- %&gt;% // or glmFit according to choice -->
		edgeR::glmQLFTest(coef = 2, contrast = my_contrasts) // or glmLRT according to choice</p>
<p>Underlying method for DESeq2 framework:
keep_abundant(
		factor_of_interest = !!as.symbol(parse_formula(.formula)[[1]]),
		minimum_counts = minimum_counts,
		minimum_proportion = minimum_proportion
) <!-- %&gt;% --></p>
<p># DESeq2
DESeq2::DESeqDataSet( design = .formula) <!-- %&gt;% -->
DESeq2::DESeq() <!-- %&gt;% -->
DESeq2::results()</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
 <span class='co'># edgeR</span>

 <span class='fu'>tidybulk</span><span class='fu'>::</span><span class='va'><a href='counts_mini.html'>counts_mini</a></span> <span class='op'>%&gt;%</span>
 <span class='fu'><a href='tidybulk.html'>tidybulk</a></span><span class='op'>(</span><span class='va'>sample</span>, <span class='va'>transcript</span>, <span class='va'>count</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
 <span class='fu'><a href='identify_abundant.html'>identify_abundant</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>test_differential_abundance</span><span class='op'>(</span> <span class='op'>~</span> <span class='va'>condition</span> <span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>No group or design set. Assuming all samples belong to one group.</span></div><div class='output co'>#&gt; <span class='message'>=====================================</span>
#&gt; <span class='message'>tidybulk says: All testing methods use raw counts, irrespective of if scale_abundance</span>
#&gt; <span class='message'>or adjust_abundance have been calculated. Therefore, it is essential to add covariates</span>
#&gt; <span class='message'>such as batch effects (if applicable) in the formula.</span>
#&gt; <span class='message'>=====================================</span></div><div class='output co'>#&gt; <span class='message'>tidybulk says: The design column names are "(Intercept), conditionTRUE"</span></div><div class='output co'>#&gt; <span class='message'>tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$edgeR`</span></div><div class='output co'>#&gt; <span style='color: #949494;'># A tibble: 2,635 x 12</span><span>
#&gt;    sample   transcript `Cell type` count time  condition .abundant  logFC logCPM
#&gt;    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>      </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>       </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span>      </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>  </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>
#&gt; </span><span style='color: #BCBCBC;'> 1</span><span> SRR1740… TNFRSF4    b_cell          6 0 d   TRUE      FALSE     </span><span style='color: #BB0000;'>NA</span><span>       </span><span style='color: #BB0000;'>NA</span><span>  
#&gt; </span><span style='color: #BCBCBC;'> 2</span><span> SRR1740… PLCH2      b_cell        926 0 d   TRUE      TRUE       5.14    10.8
#&gt; </span><span style='color: #BCBCBC;'> 3</span><span> SRR1740… PADI4      b_cell         21 0 d   TRUE      TRUE      -</span><span style='color: #BB0000;'>6.27</span><span>    11.0
#&gt; </span><span style='color: #BCBCBC;'> 4</span><span> SRR1740… PAX7       b_cell          0 0 d   TRUE      FALSE     </span><span style='color: #BB0000;'>NA</span><span>       </span><span style='color: #BB0000;'>NA</span><span>  
#&gt; </span><span style='color: #BCBCBC;'> 5</span><span> SRR1740… CDA        b_cell          1 0 d   TRUE      FALSE     </span><span style='color: #BB0000;'>NA</span><span>       </span><span style='color: #BB0000;'>NA</span><span>  
#&gt; </span><span style='color: #BCBCBC;'> 6</span><span> SRR1740… RCAN3      b_cell        905 0 d   TRUE      TRUE       2.67    11.5
#&gt; </span><span style='color: #BCBCBC;'> 7</span><span> SRR1740… SMPDL3B    b_cell          3 0 d   TRUE      FALSE     </span><span style='color: #BB0000;'>NA</span><span>       </span><span style='color: #BB0000;'>NA</span><span>  
#&gt; </span><span style='color: #BCBCBC;'> 8</span><span> SRR1740… EPB41      b_cell       </span><span style='text-decoration: underline;'>4</span><span>667 0 d   TRUE      TRUE       0.858   13.2
#&gt; </span><span style='color: #BCBCBC;'> 9</span><span> SRR1740… LCK        b_cell        436 0 d   TRUE      TRUE       6.88    12.2
#&gt; </span><span style='color: #BCBCBC;'>10</span><span> SRR1740… COL8A2     b_cell          1 0 d   TRUE      FALSE     </span><span style='color: #BB0000;'>NA</span><span>       </span><span style='color: #BB0000;'>NA</span><span>  
#&gt; </span><span style='color: #949494;'># … with 2,625 more rows, and 3 more variables: F &lt;dbl&gt;, PValue &lt;dbl&gt;,</span><span>
#&gt; </span><span style='color: #949494;'>#   FDR &lt;dbl&gt;</span><span></div><div class='input'>
  <span class='co'># The function `test_differential_abundance` operates with contrasts too</span>

 <span class='fu'>tidybulk</span><span class='fu'>::</span><span class='va'><a href='counts_mini.html'>counts_mini</a></span> <span class='op'>%&gt;%</span>
 <span class='fu'><a href='tidybulk.html'>tidybulk</a></span><span class='op'>(</span><span class='va'>sample</span>, <span class='va'>transcript</span>, <span class='va'>count</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
 <span class='fu'><a href='identify_abundant.html'>identify_abundant</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
 <span class='fu'>test_differential_abundance</span><span class='op'>(</span>
      <span class='op'>~</span> <span class='fl'>0</span> <span class='op'>+</span> <span class='va'>condition</span>,
      .contrasts <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span> <span class='st'>"conditionTRUE - conditionFALSE"</span><span class='op'>)</span>
 <span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>No group or design set. Assuming all samples belong to one group.</span></div><div class='output co'>#&gt; <span class='message'>=====================================</span>
#&gt; <span class='message'>tidybulk says: All testing methods use raw counts, irrespective of if scale_abundance</span>
#&gt; <span class='message'>or adjust_abundance have been calculated. Therefore, it is essential to add covariates</span>
#&gt; <span class='message'>such as batch effects (if applicable) in the formula.</span>
#&gt; <span class='message'>=====================================</span></div><div class='output co'>#&gt; <span class='message'>tidybulk says: The design column names are "conditionFALSE, conditionTRUE"</span></div><div class='output co'>#&gt; <span class='message'>tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$edgeR`</span></div><div class='output co'>#&gt; </span><span style='color: #949494;'># A tibble: 2,635 x 12</span><span>
#&gt;    sample     transcript `Cell type` count time  condition .abundant
#&gt;    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>      </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>      </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>       </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span>    
#&gt; </span><span style='color: #BCBCBC;'> 1</span><span> SRR1740034 TNFRSF4    b_cell          6 0 d   TRUE      FALSE    
#&gt; </span><span style='color: #BCBCBC;'> 2</span><span> SRR1740034 PLCH2      b_cell        926 0 d   TRUE      TRUE     
#&gt; </span><span style='color: #BCBCBC;'> 3</span><span> SRR1740034 PADI4      b_cell         21 0 d   TRUE      TRUE     
#&gt; </span><span style='color: #BCBCBC;'> 4</span><span> SRR1740034 PAX7       b_cell          0 0 d   TRUE      FALSE    
#&gt; </span><span style='color: #BCBCBC;'> 5</span><span> SRR1740034 CDA        b_cell          1 0 d   TRUE      FALSE    
#&gt; </span><span style='color: #BCBCBC;'> 6</span><span> SRR1740034 RCAN3      b_cell        905 0 d   TRUE      TRUE     
#&gt; </span><span style='color: #BCBCBC;'> 7</span><span> SRR1740034 SMPDL3B    b_cell          3 0 d   TRUE      FALSE    
#&gt; </span><span style='color: #BCBCBC;'> 8</span><span> SRR1740034 EPB41      b_cell       </span><span style='text-decoration: underline;'>4</span><span>667 0 d   TRUE      TRUE     
#&gt; </span><span style='color: #BCBCBC;'> 9</span><span> SRR1740034 LCK        b_cell        436 0 d   TRUE      TRUE     
#&gt; </span><span style='color: #BCBCBC;'>10</span><span> SRR1740034 COL8A2     b_cell          1 0 d   TRUE      FALSE    
#&gt; </span><span style='color: #949494;'># … with 2,625 more rows, and 5 more variables:</span><span>
#&gt; </span><span style='color: #949494;'>#   logFC___conditionTRUE - conditionFALSE &lt;dbl&gt;,</span><span>
#&gt; </span><span style='color: #949494;'>#   logCPM___conditionTRUE - conditionFALSE &lt;dbl&gt;,</span><span>
#&gt; </span><span style='color: #949494;'>#   F___conditionTRUE - conditionFALSE &lt;dbl&gt;,</span><span>
#&gt; </span><span style='color: #949494;'>#   PValue___conditionTRUE - conditionFALSE &lt;dbl&gt;,</span><span>
#&gt; </span><span style='color: #949494;'>#   FDR___conditionTRUE - conditionFALSE &lt;dbl&gt;</span><span></div><div class='input'>
 <span class='co'># DESeq2 - equivalent for limma-voom</span>

 <span class='fu'>tidybulk</span><span class='fu'>::</span><span class='va'><a href='counts_mini.html'>counts_mini</a></span> <span class='op'>%&gt;%</span>
 <span class='fu'><a href='tidybulk.html'>tidybulk</a></span><span class='op'>(</span><span class='va'>sample</span>, <span class='va'>transcript</span>, <span class='va'>count</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
 <span class='fu'><a href='identify_abundant.html'>identify_abundant</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>test_differential_abundance</span><span class='op'>(</span> <span class='op'>~</span> <span class='va'>condition</span>, method<span class='op'>=</span><span class='st'>"deseq2"</span> <span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>No group or design set. Assuming all samples belong to one group.</span></div><div class='output co'>#&gt; <span class='message'>=====================================</span>
#&gt; <span class='message'>tidybulk says: All testing methods use raw counts, irrespective of if scale_abundance</span>
#&gt; <span class='message'>or adjust_abundance have been calculated. Therefore, it is essential to add covariates</span>
#&gt; <span class='message'>such as batch effects (if applicable) in the formula.</span>
#&gt; <span class='message'>=====================================</span></div><div class='output co'>#&gt; <span class='message'>estimating size factors</span></div><div class='output co'>#&gt; <span class='message'>estimating dispersions</span></div><div class='output co'>#&gt; <span class='message'>gene-wise dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>mean-dispersion relationship</span></div><div class='output co'>#&gt; <span class='message'>-- note: fitType='parametric', but the dispersion trend was not well captured by the</span>
#&gt; <span class='message'>   function: y = a/x + b, and a local regression fit was automatically substituted.</span>
#&gt; <span class='message'>   specify fitType='local' or 'mean' to avoid this message next time.</span></div><div class='output co'>#&gt; <span class='message'>final dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>fitting model and testing</span></div><div class='output co'>#&gt; <span class='message'>tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$DESeq2`</span></div><div class='output co'>#&gt; </span><span style='color: #949494;'># A tibble: 2,635 x 13</span><span>
#&gt;    sample     transcript `Cell type` count time  condition .abundant baseMean
#&gt;    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>      </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>      </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>       </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span>        </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>
#&gt; </span><span style='color: #BCBCBC;'> 1</span><span> SRR1740034 TNFRSF4    b_cell          6 0 d   TRUE      FALSE          </span><span style='color: #BB0000;'>NA</span><span> 
#&gt; </span><span style='color: #BCBCBC;'> 2</span><span> SRR1740034 PLCH2      b_cell        926 0 d   TRUE      TRUE          725.
#&gt; </span><span style='color: #BCBCBC;'> 3</span><span> SRR1740034 PADI4      b_cell         21 0 d   TRUE      TRUE          854.
#&gt; </span><span style='color: #BCBCBC;'> 4</span><span> SRR1740034 PAX7       b_cell          0 0 d   TRUE      FALSE          </span><span style='color: #BB0000;'>NA</span><span> 
#&gt; </span><span style='color: #BCBCBC;'> 5</span><span> SRR1740034 CDA        b_cell          1 0 d   TRUE      FALSE          </span><span style='color: #BB0000;'>NA</span><span> 
#&gt; </span><span style='color: #BCBCBC;'> 6</span><span> SRR1740034 RCAN3      b_cell        905 0 d   TRUE      TRUE         </span><span style='text-decoration: underline;'>1</span><span>125.
#&gt; </span><span style='color: #BCBCBC;'> 7</span><span> SRR1740034 SMPDL3B    b_cell          3 0 d   TRUE      FALSE          </span><span style='color: #BB0000;'>NA</span><span> 
#&gt; </span><span style='color: #BCBCBC;'> 8</span><span> SRR1740034 EPB41      b_cell       </span><span style='text-decoration: underline;'>4</span><span>667 0 d   TRUE      TRUE         </span><span style='text-decoration: underline;'>3</span><span>832.
#&gt; </span><span style='color: #BCBCBC;'> 9</span><span> SRR1740034 LCK        b_cell        436 0 d   TRUE      TRUE         </span><span style='text-decoration: underline;'>1</span><span>769.
#&gt; </span><span style='color: #BCBCBC;'>10</span><span> SRR1740034 COL8A2     b_cell          1 0 d   TRUE      FALSE          </span><span style='color: #BB0000;'>NA</span><span> 
#&gt; </span><span style='color: #949494;'># … with 2,625 more rows, and 5 more variables: log2FoldChange &lt;dbl&gt;,</span><span>
#&gt; </span><span style='color: #949494;'>#   lfcSE &lt;dbl&gt;, stat &lt;dbl&gt;, pvalue &lt;dbl&gt;, padj &lt;dbl&gt;</span><span></div><div class='input'>
  <span class='co'># The function `test_differential_abundance` operates with contrasts too</span>

 <span class='fu'>tidybulk</span><span class='fu'>::</span><span class='va'><a href='counts_mini.html'>counts_mini</a></span> <span class='op'>%&gt;%</span>
 <span class='fu'><a href='tidybulk.html'>tidybulk</a></span><span class='op'>(</span><span class='va'>sample</span>, <span class='va'>transcript</span>, <span class='va'>count</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
 <span class='fu'><a href='identify_abundant.html'>identify_abundant</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
 <span class='fu'>test_differential_abundance</span><span class='op'>(</span>
      <span class='op'>~</span> <span class='fl'>0</span> <span class='op'>+</span> <span class='va'>condition</span>,
      .contrasts <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"condition"</span>, <span class='st'>"TRUE"</span>, <span class='st'>"FALSE"</span><span class='op'>)</span><span class='op'>)</span>,
      method<span class='op'>=</span><span class='st'>"deseq2"</span>
 <span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>No group or design set. Assuming all samples belong to one group.</span></div><div class='output co'>#&gt; <span class='message'>=====================================</span>
#&gt; <span class='message'>tidybulk says: All testing methods use raw counts, irrespective of if scale_abundance</span>
#&gt; <span class='message'>or adjust_abundance have been calculated. Therefore, it is essential to add covariates</span>
#&gt; <span class='message'>such as batch effects (if applicable) in the formula.</span>
#&gt; <span class='message'>=====================================</span></div><div class='output co'>#&gt; <span class='message'>estimating size factors</span></div><div class='output co'>#&gt; <span class='message'>estimating dispersions</span></div><div class='output co'>#&gt; <span class='message'>gene-wise dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>mean-dispersion relationship</span></div><div class='output co'>#&gt; <span class='message'>-- note: fitType='parametric', but the dispersion trend was not well captured by the</span>
#&gt; <span class='message'>   function: y = a/x + b, and a local regression fit was automatically substituted.</span>
#&gt; <span class='message'>   specify fitType='local' or 'mean' to avoid this message next time.</span></div><div class='output co'>#&gt; <span class='message'>final dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>fitting model and testing</span></div><div class='output co'>#&gt; <span class='message'>tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$DESeq2`</span></div><div class='output co'>#&gt; </span><span style='color: #949494;'># A tibble: 2,635 x 13</span><span>
#&gt;    sample     transcript `Cell type` count time  condition .abundant
#&gt;    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>      </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>      </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>       </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span>    
#&gt; </span><span style='color: #BCBCBC;'> 1</span><span> SRR1740034 TNFRSF4    b_cell          6 0 d   TRUE      FALSE    
#&gt; </span><span style='color: #BCBCBC;'> 2</span><span> SRR1740034 PLCH2      b_cell        926 0 d   TRUE      TRUE     
#&gt; </span><span style='color: #BCBCBC;'> 3</span><span> SRR1740034 PADI4      b_cell         21 0 d   TRUE      TRUE     
#&gt; </span><span style='color: #BCBCBC;'> 4</span><span> SRR1740034 PAX7       b_cell          0 0 d   TRUE      FALSE    
#&gt; </span><span style='color: #BCBCBC;'> 5</span><span> SRR1740034 CDA        b_cell          1 0 d   TRUE      FALSE    
#&gt; </span><span style='color: #BCBCBC;'> 6</span><span> SRR1740034 RCAN3      b_cell        905 0 d   TRUE      TRUE     
#&gt; </span><span style='color: #BCBCBC;'> 7</span><span> SRR1740034 SMPDL3B    b_cell          3 0 d   TRUE      FALSE    
#&gt; </span><span style='color: #BCBCBC;'> 8</span><span> SRR1740034 EPB41      b_cell       </span><span style='text-decoration: underline;'>4</span><span>667 0 d   TRUE      TRUE     
#&gt; </span><span style='color: #BCBCBC;'> 9</span><span> SRR1740034 LCK        b_cell        436 0 d   TRUE      TRUE     
#&gt; </span><span style='color: #BCBCBC;'>10</span><span> SRR1740034 COL8A2     b_cell          1 0 d   TRUE      FALSE    
#&gt; </span><span style='color: #949494;'># … with 2,625 more rows, and 6 more variables:</span><span>
#&gt; </span><span style='color: #949494;'>#   baseMean___condition TRUE-FALSE &lt;dbl&gt;,</span><span>
#&gt; </span><span style='color: #949494;'>#   log2FoldChange___condition TRUE-FALSE &lt;dbl&gt;,</span><span>
#&gt; </span><span style='color: #949494;'>#   lfcSE___condition TRUE-FALSE &lt;dbl&gt;, stat___condition TRUE-FALSE &lt;dbl&gt;,</span><span>
#&gt; </span><span style='color: #949494;'>#   pvalue___condition TRUE-FALSE &lt;dbl&gt;, padj___condition TRUE-FALSE &lt;dbl&gt;</span><span></div><div class='input'>
</div></span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


