<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>analyse gene rank with GSEA — test_gene_rank • tidybulk</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="analyse gene rank with GSEA — test_gene_rank" />
<meta property="og:description" content="test_gene_rank() takes as input a `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment)) and returns a `tbl` with the GSEA statistics" />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidybulk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.5.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/comparison_with_base_R.html">Comparison with base R</a>
    </li>
    <li>
      <a href="../articles/introduction.html">Overview of the tidybulk package</a>
    </li>
    <li>
      <a href="../articles/manuscript_differential_transcript_abundance.html">Manuscript code - differential feature abundance</a>
    </li>
    <li>
      <a href="../articles/manuscript_transcriptional_signatures.html">Manuscript code - transcriptional signature identification</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stemangiola/tidybulk/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>analyse gene rank with GSEA</h1>
    <small class="dont-index">Source: <a href='https://github.com/stemangiola/tidybulk/blob/master/R/methods.R'><code>R/methods.R</code></a>, <a href='https://github.com/stemangiola/tidybulk/blob/master/R/methods_SE.R'><code>R/methods_SE.R</code></a></small>
    <div class="hidden name"><code>test_gene_rank-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>test_gene_rank() takes as input a `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment)) and returns a `tbl` with the GSEA statistics</p>
    </div>

    <div class="ref-usage sourceCode"><pre class='sourceCode r'><code><span class='fu'>test_gene_rank</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.entrez</span>,
  <span class='va'>.arrange_desc</span>,
  <span class='va'>species</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  gene_sets <span class='op'>=</span> <span class='cn'>NULL</span>,
  gene_set <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span>

<span class='co'># S4 method for spec_tbl_df</span>
<span class='fu'>test_gene_rank</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.entrez</span>,
  <span class='va'>.arrange_desc</span>,
  <span class='va'>species</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  gene_sets <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"h"</span>, <span class='st'>"c1"</span>, <span class='st'>"c2"</span>, <span class='st'>"c3"</span>, <span class='st'>"c4"</span>, <span class='st'>"c5"</span>, <span class='st'>"c6"</span>, <span class='st'>"c7"</span><span class='op'>)</span>,
  gene_set <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span>

<span class='co'># S4 method for tbl_df</span>
<span class='fu'>test_gene_rank</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.entrez</span>,
  <span class='va'>.arrange_desc</span>,
  <span class='va'>species</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  gene_sets <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"h"</span>, <span class='st'>"c1"</span>, <span class='st'>"c2"</span>, <span class='st'>"c3"</span>, <span class='st'>"c4"</span>, <span class='st'>"c5"</span>, <span class='st'>"c6"</span>, <span class='st'>"c7"</span><span class='op'>)</span>,
  gene_set <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span>

<span class='co'># S4 method for tidybulk</span>
<span class='fu'>test_gene_rank</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.entrez</span>,
  <span class='va'>.arrange_desc</span>,
  <span class='va'>species</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  gene_sets <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"h"</span>, <span class='st'>"c1"</span>, <span class='st'>"c2"</span>, <span class='st'>"c3"</span>, <span class='st'>"c4"</span>, <span class='st'>"c5"</span>, <span class='st'>"c6"</span>, <span class='st'>"c7"</span><span class='op'>)</span>,
  gene_set <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span>

<span class='co'># S4 method for SummarizedExperiment</span>
<span class='fu'>test_gene_rank</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.entrez</span>,
  <span class='va'>.arrange_desc</span>,
  <span class='va'>species</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  gene_sets <span class='op'>=</span> <span class='cn'>NULL</span>,
  gene_set <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span>

<span class='co'># S4 method for RangedSummarizedExperiment</span>
<span class='fu'>test_gene_rank</span><span class='op'>(</span>
  <span class='va'>.data</span>,
  <span class='va'>.entrez</span>,
  <span class='va'>.arrange_desc</span>,
  <span class='va'>species</span>,
  .sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  gene_sets <span class='op'>=</span> <span class='cn'>NULL</span>,
  gene_set <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span></code></pre></div>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>.data</th>
      <td><p>A `tbl` (with at least three columns for sample, feature and transcript abundance) or `SummarizedExperiment` (more convenient if abstracted to tibble with library(tidySummarizedExperiment))</p></td>
    </tr>
    <tr>
      <th>.entrez</th>
      <td><p>The ENTREZ ID of the transcripts/genes</p></td>
    </tr>
    <tr>
      <th>.arrange_desc</th>
      <td><p>A column name of the column to arrange in decreasing order</p></td>
    </tr>
    <tr>
      <th>species</th>
      <td><p>A character. For example, human or mouse. MSigDB uses the latin species names (e.g., \"Mus musculus\", \"Homo sapiens\")</p></td>
    </tr>
    <tr>
      <th>.sample</th>
      <td><p>The name of the sample column</p></td>
    </tr>
    <tr>
      <th>gene_sets</th>
      <td><p>A character vector or a list. It can take one or more of the following built-in collections as a character vector: c("h", "c1", "c2", "c3", "c4", "c5", "c6", "c7", "kegg_disease", "kegg_metabolism", "kegg_signaling"), to be used with EGSEA buildIdx. c1 is human specific. Alternatively, a list of user-supplied gene sets can be provided, to be used with EGSEA buildCustomIdx. In that case, each gene set is a character vector of Entrez IDs and the names of the list are the gene set names.</p></td>
    </tr>
    <tr>
      <th>gene_set</th>
      <td><p>DEPRECATED. Use gene_sets instead.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A consistent object (to the input)</p>
<p>A `spec_tbl_df` object</p>
<p>A `tbl_df` object</p>
<p>A `tidybulk` object</p>
<p>A `SummarizedExperiment` object</p>
<p>A `RangedSummarizedExperiment` object</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><a href='https://lifecycle.r-lib.org/articles/stages.html#maturing'><img src='figures/lifecycle-maturing.svg' alt='[Maturing]' /></a></p>
<p>This wrapper execute gene enrichment analyses of the dataset using a list of transcripts and GSEA.
This wrapper uses clusterProfiler (DOI: doi.org/10.1089/omi.2011.0118) on the back-end.</p>
<p>Undelying method:
# Get gene sets signatures
msigdbr::msigdbr(species = species) <!-- %&gt;% --></p>
<p># Filter specific gene_sets  if specified. This was introduced to speed up examples executionS
when(
	!is.null(gene_sets ) ~ filter(., gs_cat <!-- %in% gene_sets ), -->
	~ (.)
) <!-- %&gt;% --></p>
<p># Execute calculation
nest(data = -gs_cat) <!-- %&gt;% -->
mutate(fit =
			 	map(
			 		data,
			 		~ 	clusterProfiler::GSEA(
			 			my_entrez_rank,
			 			TERM2GENE=.x <!-- %&gt;% select(gs_name, entrez_gene), -->
			 			pvalueCutoff = 1
			 		)</p>
<p>))</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <div class="ref-examples sourceCode"><pre class='sourceCode r'><code><span class='r-in'></span>
<span class='r-in'><span class='va'>df_entrez</span> <span class='op'>=</span> <span class='fu'>tidybulk</span><span class='fu'>::</span><span class='va'><a href='se_mini.html'>se_mini</a></span> |&gt; <span class='fu'><a href='tidybulk-methods.html'>tidybulk</a></span><span class='op'>(</span><span class='op'>)</span> |&gt; <span class='fu'><a href='https://tibble.tidyverse.org/reference/as_tibble.html'>as_tibble</a></span><span class='op'>(</span><span class='op'>)</span> |&gt; <span class='fu'><a href='symbol_to_entrez.html'>symbol_to_entrez</a></span><span class='op'>(</span> .transcript <span class='op'>=</span> <span class='va'>feature</span>, .sample <span class='op'>=</span> <span class='va'>sample</span><span class='op'>)</span></span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> 'select()' returned 1:many mapping between keys and columns</span>
<span class='r-in'><span class='va'>df_entrez</span> <span class='op'>=</span> <span class='fu'><a href='aggregate_duplicates-methods.html'>aggregate_duplicates</a></span><span class='op'>(</span><span class='va'>df_entrez</span>, aggregation_function <span class='op'>=</span> <span class='va'>sum</span>, .sample <span class='op'>=</span> <span class='va'>sample</span>, .transcript <span class='op'>=</span> <span class='va'>entrez</span>, .abundance <span class='op'>=</span> <span class='va'>count</span><span class='op'>)</span></span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>tidybulk says: for aggregation, factors and logical columns were converted to character</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Converted to characters</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> factorfactorfactorfactorlogical</span>
<span class='r-in'><span class='va'>df_entrez</span> <span class='op'>=</span> <span class='fu'><a href='mutate-methods.html'>mutate</a></span><span class='op'>(</span><span class='va'>df_entrez</span>, do_test <span class='op'>=</span> <span class='va'>feature</span> <span class='op'><a href='https://rdrr.io/r/base/match.html'>%in%</a></span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"TNFRSF4"</span>, <span class='st'>"PLCH2"</span>, <span class='st'>"PADI4"</span>, <span class='st'>"PAX7"</span><span class='op'>)</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>df_entrez</span>  <span class='op'>=</span> <span class='va'>df_entrez</span> <span class='op'><a href='pipe.html'>%&gt;%</a></span> <span class='fu'><a href='test_differential_abundance-methods.html'>test_differential_abundance</a></span><span class='op'>(</span><span class='op'>~</span> <span class='va'>condition</span><span class='op'>)</span></span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>tidybulk says: highly abundant transcripts were not identified (i.e. identify_abundant()) or filtered (i.e., keep_abundant), therefore this operation will be performed on unfiltered data. In rare occasions this could be wanted. In standard whole-transcriptome workflows is generally unwanted.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> tidybulk says: The design column names are "(Intercept), conditionTRUE"</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$edgeR`</span>
<span class='r-in'></span>
<span class='r-in'><span class='fu'>test_gene_rank</span><span class='op'>(</span></span>
<span class='r-in'>  <span class='va'>df_entrez</span>,</span>
<span class='r-in'>    .sample <span class='op'>=</span> <span class='va'>sample</span>,</span>
<span class='r-in'>  .entrez <span class='op'>=</span> <span class='va'>entrez</span>,</span>
<span class='r-in'>    species<span class='op'>=</span><span class='st'>"Homo sapiens"</span>,</span>
<span class='r-in'>   gene_sets <span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"C2"</span><span class='op'>)</span>,</span>
<span class='r-in'> .arrange_desc <span class='op'>=</span> <span class='va'>logFC</span></span>
<span class='r-in'>  <span class='op'>)</span></span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>tidybulk says: there are .entrez that are NA. Those will be removed</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> preparing geneSet collections...</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> GSEA analysis...</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>There are ties in the preranked stats (1.35% of the list).</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> The order of those tied genes will be arbitrary, which may produce unexpected results.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> leading edge analysis...</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> done...</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> <span style='color: #949494;'># A tibble: 1 × 3</span></span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   gs_cat fit             test               </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>   <span style='color: #949494; font-style: italic;'>&lt;chr&gt;</span>  <span style='color: #949494; font-style: italic;'>&lt;list&gt;</span>          <span style='color: #949494; font-style: italic;'>&lt;list&gt;</span>             </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> <span style='color: #BCBCBC;'>1</span> C2     <span style='color: #949494;'>&lt;gseaRslt[,11]&gt;</span> <span style='color: #949494;'>&lt;tibble [498 × 15]&gt;</span></span>
<span class='r-in'></span>
<span class='r-in'></span>
</code></pre></div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by Stefano Mangiola.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.9000.9000.9000.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


