my_bibliography = function() {
	list(
	tidybulk = " @Article{tidybulk,
  title = {tidybulk: an R tidy framework for modular transcriptomic data analysis},
  author = {Stefano Mangiola and Ramyar Molania and Ruining Dong and Maria A. Doyle & Anthony T. Papenfuss},
  journal = {Genome Biology},
  year = {2021},
  volume = {22},
  number = {42},
  url = {https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02233-7},
  }",
	tidyverse = "@article{wickham2019welcome,
  title={Welcome to the Tidyverse},
  author={Wickham, Hadley and Averick, Mara and Bryan, Jennifer and Chang, Winston and McGowan, Lucy D'Agostino and Francois, Romain and Grolemund, Garrett and Hayes, Alex and Henry, Lionel and Hester, Jim and others},
  journal={Journal of Open Source Software},
  volume={4},
  number={43},
  pages={1686},
  year={2019}
	}",
	featurecounts = "@article{liao2014featurecounts,
  title={featureCounts: an efficient general purpose program for assigning sequence reads to genomic features},
  author={Liao, Yang and Smyth, Gordon K and Shi, Wei},
  journal={Bioinformatics},
  volume={30},
  number={7},
  pages={923--930},
  year={2014},
  publisher={Oxford University Press}
	}" ,
	edger = "@article{robinson2010edger,
  title={edgeR: a Bioconductor package for differential expression analysis of digital gene expression data},
  author={Robinson, Mark D and McCarthy, Davis J and Smyth, Gordon K},
  journal={Bioinformatics},
  volume={26},
  number={1},
  pages={139--140},
  year={2010},
  publisher={Oxford University Press}
	}",
	edgeR_quasi_likelihood = "@article{lund2012detecting,
  title={Detecting differential expression in RNA-sequence data using quasi-likelihood with shrunken dispersion estimates},
  author={Lund, Steven P and Nettleton, Dan and McCarthy, Davis J and Smyth, Gordon K},
  journal={Statistical applications in genetics and molecular biology},
  volume={11},
  number={5},
  year={2012},
  publisher={De Gruyter}
    }",
	edgeR_likelihood_ratio = "@article{mccarthy2012differential,
  title={Differential expression analysis of multifactor RNA-Seq experiments with respect to biological variation},
  author={McCarthy, Davis J and Chen, Yunshun and Smyth, Gordon K},
  journal={Nucleic acids research},
  volume={40},
  number={10},
  pages={4288--4297},
  year={2012},
  publisher={Oxford University Press}
    }",
	edger_robust_likelihood_ratio = "@article{zhou2014robustly,
  title={Robustly detecting differential expression in RNA sequencing data using observation weights},
  author={Zhou, Xiaobei and Lindsay, Helen and Robinson, Mark D},
  journal={Nucleic acids research},
  volume={42},
  number={11},
  pages={e91--e91},
  year={2014},
  publisher={Oxford University Press}
	}",
	voom = "@article{law2014voom,
  title={voom: Precision weights unlock linear model analysis tools for RNA-seq read counts},
  author={Law, Charity W and Chen, Yunshun and Shi, Wei and Smyth, Gordon K},
  journal={Genome biology},
  volume={15},
  number={2},
  pages={R29},
  year={2014},
  publisher={Springer}
    }",
	tmm = "@article{robinson2010scaling,
  title={A scaling normalization method for differential expression analysis of RNA-seq data},
  author={Robinson, Mark D and Oshlack, Alicia},
  journal={Genome biology},
  volume={11},
  number={3},
  pages={1--9},
  year={2010},
  publisher={BioMed Central}
	}",
	limma = "@incollection{smyth2005limma,
  title={Limma: linear models for microarray data},
  author={Smyth, Gordon K},
  booktitle={Bioinformatics and computational biology solutions using R and Bioconductor},
  pages={397--420},
  year={2005},
  publisher={Springer}
	}",
	voom_sample_weights = "@article{liu2015weight,
  title={Why weight? Modelling sample and observational level variability improves power in RNA-seq analyses},
  author={Liu, Ruijie and Holik, Aliaksei Z and Su, Shian and Jansz, Natasha and Chen, Kelan and Leong, Huei San and Blewitt, Marnie E and Asselin-Labat, Marie-Liesse and Smyth, Gordon K and Ritchie, Matthew E},
  journal={Nucleic acids research},
  volume={43},
  number={15},
  pages={e97--e97},
  year={2015},
  publisher={Oxford University Press}
    }",
	deseq2 = "@article{love2014moderated,
  title={Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2},
  author={Love, Michael I and Huber, Wolfgang and Anders, Simon},
  journal={Genome biology},
  volume={15},
  number={12},
  pages={550},
  year={2014},
  publisher={Springer}
	}",
	survival = "@book{Grambsch2000-oo,
	  title   = {Modeling survival data: extending the Cox model},
	  author  = {Grambsch, Patricia M and Therneau, Terry M},
	  journal = {Stat. Biol. Health},
	  year    =  2000
	}",
	boot = "@book{davison1997bootstrap,
  title={Bootstrap methods and their application},
  author={Davison, Anthony Christopher and Hinkley, David Victor},
  number={1},
  year={1997},
  publisher={Cambridge university press}
	}",
	betareg = "@article{cribari2009beta,
  title={Beta regression in R},
  author={Cribari-Neto, Francisco and Zeileis, Achim},
  year={2009},
  publisher={Department of Statistics and Mathematics x, WU Vienna University}
	}",
	stats = "@Manual{,
    title = {R: A Language and Environment for Statistical Computing},
    author = {{R Core Team}},
    organization = {R Foundation for Statistical Computing},
    address = {Vienna, Austria},
    year = {2020},
    url = {https://www.R-project.org/},
  }",
	seurat = "@article{Butler2018-ae,
  title    = {Integrating single-cell transcriptomic data across different
	conditions, technologies, and species},
  author   = {Butler, Andrew and Hoffman, Paul and Smibert, Peter and Papalexi,
	Efthymia and Satija, Rahul},
  journal  = {Nat. Biotechnol.},
  volume   =  36,
  number   =  5,
  pages    = {411--420},
  month    =  jun,
  year     =  2018,
  language = {en}
	}",
	rtsne = "  @Article{,
    title = {Visualizing High-Dimensional Data Using t-SNE},
    volume = {9},
    pages = {2579-2605},
    year = {2008},
    author = {L.J.P. {van der Maaten} and G.E. Hinton},
    journal = {Journal of Machine Learning Research},
  }",
	widyr ="@Manual{,
    title = {widyr: Widen, Process, then Re-Tidy Data},
    author = {David Robinson},
    year = {2020},
    note = {R package version 0.1.3},
    url = {https://CRAN.R-project.org/package=widyr},
  }",
	cibersort = "@article{newman2015robust,
  title={Robust enumeration of cell subsets from tissue expression profiles},
  author={Newman, Aaron M and Liu, Chih Long and Green, Michael R and Gentles, Andrew J and Feng, Weiguo and Xu, Yue and Hoang, Chuong D and Diehn, Maximilian and Alizadeh, Ash A},
  journal={Nature methods},
  volume={12},
  number={5},
  pages={453--457},
  year={2015},
  publisher={Nature Publishing Group}
	}",
	epic = "@article{racle2017simultaneous,
  title={Simultaneous enumeration of cancer and immune cell types from bulk tumor gene expression data},
  author={Racle, Julien and de Jonge, Kaat and Baumgaertner, Petra and Speiser, Daniel E and Gfeller, David},
  journal={Elife},
  volume={6},
  pages={e26476},
  year={2017},
  publisher={eLife Sciences Publications Limited}
    }",
	llsr = "@article{abbas2009deconvolution,
  title={Deconvolution of blood microarray data identifies cellular activation patterns in systemic lupus erythematosus},
  author={Abbas, Alexander R and Wolslegel, Kristen and Seshasayee, Dhaya and Modrusan, Zora and Clark, Hilary F},
  journal={PloS one},
  volume={4},
  number={7},
  pages={e6098},
  year={2009},
  publisher={Public Library of Science}
    }",
	sva = "@article{leek2012sva,
  title={The sva package for removing batch effects and other unwanted variation in high-throughput experiments},
  author={Leek, Jeffrey T and Johnson, W Evan and Parker, Hilary S and Jaffe, Andrew E and Storey, John D},
  journal={Bioinformatics},
  volume={28},
  number={6},
  pages={882--883},
  year={2012},
  publisher={Oxford University Press}
	}",
	clusterProfiler = "@Article{,
    title = {clusterProfiler: an R package for comparing biological themes among gene clusters},
    author = {Guangchuang Yu and Li-Gen Wang and Yanyan Han and Qing-Yu He},
    journal = {OMICS: A Journal of Integrative Biology},
    year = {2012},
    volume = {16},
    number = {5},
    pages = {284-287},
    pmid = {22455463},
    doi = {10.1089/omi.2011.0118},
  }",
	msigdbr = "@Manual{,
    title = {msigdbr: MSigDB Gene Sets for Multiple Organisms in a Tidy Data Format},
    author = {Igor Dolgalev},
    year = {2020},
    note = {R package version 7.1.1},
    url = {https://CRAN.R-project.org/package=msigdbr},
  }",
	msigdb = "@article{liberzon2011molecular,
  title={Molecular signatures database (MSigDB) 3.0},
  author={Liberzon, Arthur and Subramanian, Aravind and Pinchback, Reid and Thorvaldsdottir, Helga and Tamayo, Pablo and Mesirov, Jill P},
  journal={Bioinformatics},
  volume={27},
  number={12},
  pages={1739--1740},
  year={2011},
  publisher={Oxford University Press}
	}",
	egsea = "@article{alhamdoosh2017combining,
  title={Combining multiple tools outperforms individual methods in gene set enrichment analyses},
  author={Alhamdoosh, Monther and Ng, Milica and Wilson, Nicholas J and Sheridan, Julie M and Huynh, Huy and Wilson, Michael J and Ritchie, Matthew E},
  journal={Bioinformatics},
  volume={33},
  number={3},
  pages={414--424},
  year={2017},
  publisher={Oxford University Press}
  }",
	camera = "@article{Wu2012,
  doi = {10.1093/nar/gks461},
  url = {https://doi.org/10.1093/nar/gks461},
  year = {2012},
  month = may,
  publisher = {Oxford University Press ({OUP})},
  volume = {40},
  number = {17},
  pages = {e133--e133},
  author = {Di Wu and Gordon K. Smyth},
  title = {Camera: a competitive gene set test accounting for inter-gene correlation},
  journal = {Nucleic Acids Research}
}",
    roast = "@article{Wu2010,
  doi = {10.1093/bioinformatics/btq401},
  url = {https://doi.org/10.1093/bioinformatics/btq401},
  year = {2010},
  month = jul,
  publisher = {Oxford University Press ({OUP})},
  volume = {26},
  number = {17},
  pages = {2176--2182},
  author = {Di Wu and Elgene Lim and Fran\u0037ois Vaillant and Marie-Liesse Asselin-Labat and Jane E. Visvader and Gordon K. Smyth},
  title = {{ROAST}: rotation gene set tests for complex microarray experiments},
  journal = {Bioinformatics}
}",
    fry ="@article{giner2016fry,
  title={FRY: a fast approximation to ROAST gene set test with mean aggregated set statistics},
  author={Giner, G{\"o}knur and Smyth, Gordon K},
  journal={F1000Research},
  volume={5},
  year={2016}
}",
    safe = "@article{barry2005significance,
  title={Significance analysis of functional categories in gene expression studies: a structured permutation approach},
  author={Barry, William T and Nobel, Andrew B and Wright, Fred A},
  journal={Bioinformatics},
  volume={21},
  number={9},
  pages={1943--1949},
  year={2005},
  publisher={Oxford University Press}
}",
    gage = "@article{luo2009gage,
  title={GAGE: generally applicable gene set enrichment for pathway analysis},
  author={Luo, Weijun and Friedman, Michael S and Shedden, Kerby and Hankenson, Kurt D and Woolf, Peter J},
  journal={BMC bioinformatics},
  volume={10},
  number={1},
  pages={1--17},
  year={2009},
  publisher={Springer}
}",
    padog = "@article{tarca2012down,
  title={Down-weighting overlapping genes improves gene set analysis},
  author={Tarca, Adi Laurentiu and Draghici, Sorin and Bhatti, Gaurav and Romero, Roberto},
  journal={BMC bioinformatics},
  volume={13},
  number={1},
  pages={1--14},
  year={2012},
  publisher={BioMed Central}
}",
    plage = "@article{tomfohr2005pathway,
  title={Pathway level analysis of gene expression using singular value decomposition},
  author={Tomfohr, John and Lu, Jun and Kepler, Thomas B},
  journal={BMC bioinformatics},
  volume={6},
  number={1},
  pages={1--11},
  year={2005},
  publisher={BioMed Central}
}",
    zscore = "@article{lee2008inferring,
  title={Inferring pathway activity toward precise disease classification},
  author={Lee, Eunjung and Chuang, Han-Yu and Kim, Jong-Won and Ideker, Trey and Lee, Doheon},
  journal={PLoS comput biol},
  volume={4},
  number={11},
  pages={e1000217},
  year={2008},
  publisher={Public Library of Science}
}",
    gsva = "@misc{hanzelmann2013gsva,
  title={GSVA: The Gene Set Variation Analysis package for microarray and RNA-seq data},
  author={H{\"a}nzelmann, Sonja and Castelo, Robert and Guinney, Justin},
  year={2013},
  publisher={vol}
}",
    ssgsea = "@article{barbie2009systematic,
  title={Systematic RNA interference reveals that oncogenic KRAS-driven cancers require TBK1},
  author={Barbie, David A and Tamayo, Pablo and Boehm, Jesse S and Kim, So Young and Moody, Susan E and Dunn, Ian F and Schinzel, Anna C and Sandy, Peter and Meylan, Etienne and Scholl, Claudia and others},
  journal={Nature},
  volume={462},
  number={7269},
  pages={108--112},
  year={2009},
  publisher={Nature Publishing Group}
}",
    globaltest = "@article{goeman2004global,
  title={A global test for groups of genes: testing association with a clinical outcome},
  author={Goeman, Jelle J and Van De Geer, Sara A and De Kort, Floor and Van Houwelingen, Hans C},
  journal={Bioinformatics},
  volume={20},
  number={1},
  pages={93--99},
  year={2004},
  publisher={Oxford University Press}
}",
    ora = "@article{tavazoie1999systematic,
  title={Systematic determination of genetic network architecture},
  author={Tavazoie, Saeed and Hughes, Jason D and Campbell, Michael J and Cho, Raymond J and Church, George M},
  journal={Nature genetics},
  volume={22},
  number={3},
  pages={281--285},
  year={1999},
  publisher={Nature Publishing Group}
}",
    kegg = "@article{Kanehisa2000,
  doi = {10.1093/nar/28.1.27},
  url = {https://doi.org/10.1093/nar/28.1.27},
  year = {2000},
  month = jan,
  publisher = {Oxford University Press ({OUP})},
  volume = {28},
  number = {1},
  pages = {27--30},
  author = {M. Kanehisa},
  title = {{KEGG}: Kyoto Encyclopedia of Genes and Genomes},
  journal = {Nucleic Acids Research}
}",
	survminer = c( "@Manual{,
    title = {survminer: Drawing Survival Curves using 'ggplot2'},
    author = {Alboukadel Kassambara and Marcin Kosinski and Przemyslaw Biecek},
    year = {2020},
    note = {R package version 0.4.8},
    url = {https://CRAN.R-project.org/package=survminer},
  }")
)
}
